{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMQZBoBKVpxA7gHiOWkuewC",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/vdhulappanavar/bioinformatics/blob/main/project.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 382
        },
        "id": "jxvOosvKzAKs",
        "outputId": "7317f86b-308a-4960-ba2b-32bdc4bede21"
      },
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "'<' not supported between instances of 'str' and 'int'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-3-f8f318799048>\u001b[0m in \u001b[0;36m<cell line: 48>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[0;31m# Random Forest Model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m \u001b[0mrf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRandomForestClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_estimators\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m \u001b[0mrf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;31m# Prediction and Evaluation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1150\u001b[0m                 )\n\u001b[1;32m   1151\u001b[0m             ):\n\u001b[0;32m-> 1152\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfit_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mestimator\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1154\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[1;32m    388\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_outputs_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    389\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 390\u001b[0;31m         \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpanded_class_weight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_y_class_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    392\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mDOUBLE\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontiguous\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36m_validate_y_class_weight\u001b[0;34m(self, y)\u001b[0m\n\u001b[1;32m    747\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    748\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate_y_class_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 749\u001b[0;31m         \u001b[0mcheck_classification_targets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    750\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    751\u001b[0m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mcheck_classification_targets\u001b[0;34m(y)\u001b[0m\n\u001b[1;32m    206\u001b[0m         \u001b[0mTarget\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m     \"\"\"\n\u001b[0;32m--> 208\u001b[0;31m     \u001b[0my_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minput_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"y\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    209\u001b[0m     if y_type not in [\n\u001b[1;32m    210\u001b[0m         \u001b[0;34m\"binary\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mtype_of_target\u001b[0;34m(y, input_name)\u001b[0m\n\u001b[1;32m    386\u001b[0m     \u001b[0;31m# Check multiclass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    387\u001b[0m     \u001b[0mfirst_row\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0missparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetrow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 388\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mxp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfirst_row\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    389\u001b[0m         \u001b[0;31m# [1, 2, 3] or [[1., 2., 3]] or [[1, 2]]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    390\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m\"multiclass\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msuffix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_array_api.py\u001b[0m in \u001b[0;36munique_values\u001b[0;34m(self, x)\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0munique_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/arraysetops.py\u001b[0m in \u001b[0;36munique\u001b[0;34m(ar, return_index, return_inverse, return_counts, axis, equal_nan)\u001b[0m\n\u001b[1;32m    272\u001b[0m     \u001b[0mar\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masanyarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mar\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    273\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 274\u001b[0;31m         ret = _unique1d(ar, return_index, return_inverse, return_counts, \n\u001b[0m\u001b[1;32m    275\u001b[0m                         equal_nan=equal_nan)\n\u001b[1;32m    276\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_unpack_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/arraysetops.py\u001b[0m in \u001b[0;36m_unique1d\u001b[0;34m(ar, return_index, return_inverse, return_counts, equal_nan)\u001b[0m\n\u001b[1;32m    334\u001b[0m         \u001b[0maux\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mperm\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 336\u001b[0;31m         \u001b[0mar\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    337\u001b[0m         \u001b[0maux\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mar\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m     \u001b[0mmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maux\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbool_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: '<' not supported between instances of 'str' and 'int'"
          ]
        }
      ],
      "source": [
        "import pandas as pd\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "\n",
        "# Load the data\n",
        "pdb_data_no_dups = pd.read_csv('./pdb_data_no_dups.csv')\n",
        "pdb_data_seq = pd.read_csv('./pdb_data_seq.csv')\n",
        "\n",
        "# Merge the datasets on 'structureId'\n",
        "merged_data = pd.merge(pdb_data_no_dups, pdb_data_seq, on='structureId')\n",
        "\n",
        "# Ensure 'sequence' column is of type str and handle missing values\n",
        "merged_data['sequence'] = merged_data['sequence'].fillna('').astype(str)\n",
        "\n",
        "# Feature Extraction: Create 'sequence_length' by calculating the length of each sequence\n",
        "if 'sequence' in merged_data.columns:\n",
        "    merged_data['sequence_length'] = merged_data['sequence'].apply(len)\n",
        "else:\n",
        "    raise KeyError(\"The 'sequence' column is missing from the dataset\")\n",
        "\n",
        "# Convert 'residueCount_x' to numeric (this assumes 'residueCount_x' is the correct column name)\n",
        "merged_data['residueCount_x'] = pd.to_numeric(merged_data['residueCount_x'], errors='coerce')\n",
        "\n",
        "# Handle any remaining NaN values in numeric columns\n",
        "merged_data.fillna(0, inplace=True)\n",
        "\n",
        "# Example: Calculate amino acid composition\n",
        "def amino_acid_composition(sequence):\n",
        "    aa_dict = {aa: sequence.count(aa) for aa in set(sequence)}\n",
        "    return aa_dict\n",
        "\n",
        "merged_data['aa_composition'] = merged_data['sequence'].apply(amino_acid_composition)\n",
        "\n",
        "# Feature Scaling\n",
        "scaler = StandardScaler()\n",
        "X = scaler.fit_transform(merged_data[['sequence_length', 'residueCount_x']])\n",
        "\n",
        "# Target Variable\n",
        "y = merged_data['classification']\n",
        "\n",
        "# Train-Test Split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Random Forest Model\n",
        "rf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
        "rf.fit(X_train, y_train)\n",
        "\n",
        "# Prediction and Evaluation\n",
        "y_pred = rf.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
        "print(classification_report(y_test, y_pred))\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check the data types of all columns\n",
        "print(merged_data.dtypes)\n",
        "\n",
        "# Optionally, inspect the first few rows to get an overview\n",
        "print(merged_data.head())\n",
        "\n",
        "# Identify non-numeric entries in the columns of interest\n",
        "for col in ['sequence_length', 'residueCount_x']:\n",
        "    non_numeric = merged_data[pd.to_numeric(merged_data[col], errors='coerce').isna()]\n",
        "    if not non_numeric.empty:\n",
        "        print(f\"Non-numeric values found in column {col}:\")\n",
        "        print(non_numeric[[col]])\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ueDvZDQizrpD",
        "outputId": "212cea4e-52c3-4002-e477-7dc38d2e7656"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "structureId                  object\n",
            "classification               object\n",
            "experimentalTechnique        object\n",
            "macromoleculeType_x          object\n",
            "residueCount_x                int64\n",
            "resolution                  float64\n",
            "structureMolecularWeight    float64\n",
            "crystallizationMethod        object\n",
            "crystallizationTempK        float64\n",
            "densityMatthews             float64\n",
            "densityPercentSol           float64\n",
            "pdbxDetails                  object\n",
            "phValue                     float64\n",
            "publicationYear             float64\n",
            "chainId                      object\n",
            "sequence                     object\n",
            "residueCount_y                int64\n",
            "macromoleculeType_y          object\n",
            "sequence_length               int64\n",
            "aa_composition               object\n",
            "dtype: object\n",
            "  structureId    classification experimentalTechnique macromoleculeType_x  \\\n",
            "0        100D    DNA-RNA HYBRID     X-RAY DIFFRACTION      DNA/RNA Hybrid   \n",
            "1        100D    DNA-RNA HYBRID     X-RAY DIFFRACTION      DNA/RNA Hybrid   \n",
            "2        101D               DNA     X-RAY DIFFRACTION                 DNA   \n",
            "3        101D               DNA     X-RAY DIFFRACTION                 DNA   \n",
            "4        101M  OXYGEN TRANSPORT     X-RAY DIFFRACTION             Protein   \n",
            "\n",
            "   residueCount_x  resolution  structureMolecularWeight  \\\n",
            "0              20        1.90                   6360.30   \n",
            "1              20        1.90                   6360.30   \n",
            "2              24        2.25                   7939.35   \n",
            "3              24        2.25                   7939.35   \n",
            "4             154        2.07                  18112.80   \n",
            "\n",
            "           crystallizationMethod  crystallizationTempK  densityMatthews  \\\n",
            "0  VAPOR DIFFUSION, HANGING DROP                   0.0             1.78   \n",
            "1  VAPOR DIFFUSION, HANGING DROP                   0.0             1.78   \n",
            "2                              0                   0.0             2.00   \n",
            "3                              0                   0.0             2.00   \n",
            "4                              0                   0.0             3.09   \n",
            "\n",
            "   densityPercentSol                                        pdbxDetails  \\\n",
            "0              30.89             pH 7.00, VAPOR DIFFUSION, HANGING DROP   \n",
            "1              30.89             pH 7.00, VAPOR DIFFUSION, HANGING DROP   \n",
            "2              38.45                                                  0   \n",
            "3              38.45                                                  0   \n",
            "4              60.20  3.0 M AMMONIUM SULFATE, 20 MM TRIS, 1MM EDTA, ...   \n",
            "\n",
            "   phValue  publicationYear chainId  \\\n",
            "0      7.0           1994.0       A   \n",
            "1      7.0           1994.0       B   \n",
            "2      0.0           1995.0       A   \n",
            "3      0.0           1995.0       B   \n",
            "4      9.0           1999.0       A   \n",
            "\n",
            "                                            sequence  residueCount_y  \\\n",
            "0                                         CCGGCGCCGG              20   \n",
            "1                                         CCGGCGCCGG              20   \n",
            "2                                       CGCGAATTCGCG              24   \n",
            "3                                       CGCGAATTCGCG              24   \n",
            "4  MVLSEGEWQLVLHVWAKVEADVAGHGQDILIRLFKSHPETLEKFDR...             154   \n",
            "\n",
            "  macromoleculeType_y  sequence_length  \\\n",
            "0      DNA/RNA Hybrid               10   \n",
            "1      DNA/RNA Hybrid               10   \n",
            "2                 DNA               12   \n",
            "3                 DNA               12   \n",
            "4             Protein              154   \n",
            "\n",
            "                                      aa_composition  \n",
            "0                                   {'C': 5, 'G': 5}  \n",
            "1                                   {'C': 5, 'G': 5}  \n",
            "2                   {'C': 4, 'T': 2, 'A': 2, 'G': 4}  \n",
            "3                   {'C': 4, 'T': 2, 'A': 2, 'G': 4}  \n",
            "4  {'D': 6, 'F': 5, 'W': 2, 'T': 5, 'K': 19, 'E':...  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Convert the columns to numeric, forcing any non-convertible values to NaN\n",
        "merged_data['residueCount_x'] = pd.to_numeric(merged_data['residueCount_x'], errors='coerce')\n",
        "merged_data['sequence_length'] = pd.to_numeric(merged_data['sequence_length'], errors='coerce')\n",
        "\n",
        "# Handle NaN values after conversion\n",
        "merged_data.fillna(0, inplace=True)\n",
        "\n",
        "# Now proceed with the rest of the analysis\n"
      ],
      "metadata": {
        "id": "AutMQEQMzuat"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "\n",
        "# Feature Scaling\n",
        "scaler = StandardScaler()\n",
        "X = scaler.fit_transform(merged_data[['sequence_length', 'residueCount_x']])\n",
        "\n",
        "# Target Variable\n",
        "y = merged_data['classification']\n",
        "\n",
        "# Train-Test Split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Random Forest Model\n",
        "rf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
        "rf.fit(X_train, y_train)\n",
        "\n",
        "# Prediction and Evaluation\n",
        "y_pred = rf.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
        "print(classification_report(y_test, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 382
        },
        "id": "SvuSaWSTzwJf",
        "outputId": "5763596c-f5ca-4074-9d9c-b8b887b38c1a"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "'<' not supported between instances of 'str' and 'int'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-13-54be5c1d885d>\u001b[0m in \u001b[0;36m<cell line: 18>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;31m# Random Forest Model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0mrf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRandomForestClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_estimators\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m \u001b[0mrf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;31m# Prediction and Evaluation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1150\u001b[0m                 )\n\u001b[1;32m   1151\u001b[0m             ):\n\u001b[0;32m-> 1152\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfit_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mestimator\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1154\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[1;32m    388\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_outputs_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    389\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 390\u001b[0;31m         \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpanded_class_weight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_y_class_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    392\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mDOUBLE\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontiguous\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36m_validate_y_class_weight\u001b[0;34m(self, y)\u001b[0m\n\u001b[1;32m    747\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    748\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate_y_class_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 749\u001b[0;31m         \u001b[0mcheck_classification_targets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    750\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    751\u001b[0m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mcheck_classification_targets\u001b[0;34m(y)\u001b[0m\n\u001b[1;32m    206\u001b[0m         \u001b[0mTarget\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m     \"\"\"\n\u001b[0;32m--> 208\u001b[0;31m     \u001b[0my_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minput_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"y\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    209\u001b[0m     if y_type not in [\n\u001b[1;32m    210\u001b[0m         \u001b[0;34m\"binary\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mtype_of_target\u001b[0;34m(y, input_name)\u001b[0m\n\u001b[1;32m    386\u001b[0m     \u001b[0;31m# Check multiclass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    387\u001b[0m     \u001b[0mfirst_row\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0missparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetrow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 388\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mxp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfirst_row\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    389\u001b[0m         \u001b[0;31m# [1, 2, 3] or [[1., 2., 3]] or [[1, 2]]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    390\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m\"multiclass\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msuffix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_array_api.py\u001b[0m in \u001b[0;36munique_values\u001b[0;34m(self, x)\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0munique_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/arraysetops.py\u001b[0m in \u001b[0;36munique\u001b[0;34m(ar, return_index, return_inverse, return_counts, axis, equal_nan)\u001b[0m\n\u001b[1;32m    272\u001b[0m     \u001b[0mar\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masanyarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mar\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    273\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 274\u001b[0;31m         ret = _unique1d(ar, return_index, return_inverse, return_counts, \n\u001b[0m\u001b[1;32m    275\u001b[0m                         equal_nan=equal_nan)\n\u001b[1;32m    276\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_unpack_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/arraysetops.py\u001b[0m in \u001b[0;36m_unique1d\u001b[0;34m(ar, return_index, return_inverse, return_counts, equal_nan)\u001b[0m\n\u001b[1;32m    334\u001b[0m         \u001b[0maux\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mperm\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 336\u001b[0;31m         \u001b[0mar\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    337\u001b[0m         \u001b[0maux\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mar\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m     \u001b[0mmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maux\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbool_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: '<' not supported between instances of 'str' and 'int'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check for unexpected string values in numeric columns\n",
        "# Convert any placeholders to NaN, then handle them\n",
        "\n",
        "columns_to_check = ['phValue', 'crystallizationTempK', 'densityMatthews', 'densityPercentSol']\n",
        "\n",
        "for col in columns_to_check:\n",
        "    merged_data[col] = pd.to_numeric(merged_data[col], errors='coerce')\n",
        "\n",
        "# Handle NaN values that arise after coercion\n",
        "merged_data.fillna(0, inplace=True)\n",
        "\n",
        "# Now proceed with the rest of the analysis\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "\n",
        "# Feature Scaling\n",
        "scaler = StandardScaler()\n",
        "X = scaler.fit_transform(merged_data[['sequence_length', 'residueCount_x']])\n",
        "\n",
        "# Target Variable\n",
        "y = merged_data['classification']\n",
        "\n",
        "# Train-Test Split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Random Forest Model\n",
        "rf = RandomForestClassifier(n_estimators=100, random_state=42)\n",
        "rf.fit(X_train, y_train)\n",
        "\n",
        "# Prediction and Evaluation\n",
        "y_pred = rf.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
        "print(classification_report(y_test, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 382
        },
        "id": "WXsAmryHzzTf",
        "outputId": "15993dc7-4e87-4f12-9014-591bcb5febfa"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "'<' not supported between instances of 'str' and 'int'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-6-6fd49d9ad5a4>\u001b[0m in \u001b[0;36m<cell line: 30>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;31m# Random Forest Model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0mrf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRandomForestClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_estimators\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 30\u001b[0;31m \u001b[0mrf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;31m# Prediction and Evaluation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1150\u001b[0m                 )\n\u001b[1;32m   1151\u001b[0m             ):\n\u001b[0;32m-> 1152\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfit_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mestimator\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1154\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[1;32m    388\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_outputs_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    389\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 390\u001b[0;31m         \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpanded_class_weight\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_y_class_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    392\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mDOUBLE\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflags\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontiguous\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py\u001b[0m in \u001b[0;36m_validate_y_class_weight\u001b[0;34m(self, y)\u001b[0m\n\u001b[1;32m    747\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    748\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate_y_class_weight\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 749\u001b[0;31m         \u001b[0mcheck_classification_targets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    750\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    751\u001b[0m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mcheck_classification_targets\u001b[0;34m(y)\u001b[0m\n\u001b[1;32m    206\u001b[0m         \u001b[0mTarget\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m     \"\"\"\n\u001b[0;32m--> 208\u001b[0;31m     \u001b[0my_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minput_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"y\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    209\u001b[0m     if y_type not in [\n\u001b[1;32m    210\u001b[0m         \u001b[0;34m\"binary\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mtype_of_target\u001b[0;34m(y, input_name)\u001b[0m\n\u001b[1;32m    386\u001b[0m     \u001b[0;31m# Check multiclass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    387\u001b[0m     \u001b[0mfirst_row\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0missparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetrow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 388\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mxp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfirst_row\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    389\u001b[0m         \u001b[0;31m# [1, 2, 3] or [[1., 2., 3]] or [[1, 2]]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    390\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m\"multiclass\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msuffix\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_array_api.py\u001b[0m in \u001b[0;36munique_values\u001b[0;34m(self, x)\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0munique_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mnumpy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mconcat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marrays\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/arraysetops.py\u001b[0m in \u001b[0;36munique\u001b[0;34m(ar, return_index, return_inverse, return_counts, axis, equal_nan)\u001b[0m\n\u001b[1;32m    272\u001b[0m     \u001b[0mar\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masanyarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mar\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    273\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 274\u001b[0;31m         ret = _unique1d(ar, return_index, return_inverse, return_counts, \n\u001b[0m\u001b[1;32m    275\u001b[0m                         equal_nan=equal_nan)\n\u001b[1;32m    276\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_unpack_tuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/numpy/lib/arraysetops.py\u001b[0m in \u001b[0;36m_unique1d\u001b[0;34m(ar, return_index, return_inverse, return_counts, equal_nan)\u001b[0m\n\u001b[1;32m    334\u001b[0m         \u001b[0maux\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mar\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mperm\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    335\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 336\u001b[0;31m         \u001b[0mar\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    337\u001b[0m         \u001b[0maux\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mar\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m     \u001b[0mmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maux\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbool_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: '<' not supported between instances of 'str' and 'int'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Inspect unique values in the target variable\n",
        "print(y_train.unique())\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yJdOzRBLz2RL",
        "outputId": "3e2d5471-6f79-4455-b348-1c162fe79721"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['OXIDOREDUCTASE' 'SIGNALING PROTEIN' 'MEMBRANE PROTEIN, LIPID TRANSPORT'\n",
            " ... 'CHOLESTEROL-BINDING PROTEIN' 'VIRAL PROTEIN/WINGED HELIX'\n",
            " 'FUNGICIDE']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from sklearn.preprocessing import LabelEncoder\n",
        "\n",
        "# Encoding categorical target variable\n",
        "le = LabelEncoder()\n",
        "y_encoded = le.fit_transform(y_train)\n",
        "\n",
        "# Fit the model using encoded target variable\n",
        "rf.fit(X_train, y_encoded)\n",
        "\n",
        "# If using y_test for evaluation, ensure it is also encoded the same way\n",
        "y_test_encoded = le.transform(y_test)\n",
        "\n",
        "# Prediction and Evaluation\n",
        "y_pred = rf.predict(X_test)\n",
        "y_pred_decoded = le.inverse_transform(y_pred)  # Decode the predictions back to original labels\n",
        "print(\"Accuracy:\", accuracy_score(y_test_encoded, y_pred))\n",
        "print(classification_report(y_test_encoded, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 418
        },
        "id": "0ShYHgxHz4NB",
        "outputId": "cd09bbee-19e5-47a7-e8f3-b59ec7af3520"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "Encoders require their input argument must be uniformly strings or numbers. Got ['int', 'str']",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m_unique_python\u001b[0;34m(values, return_inverse, return_counts)\u001b[0m\n\u001b[1;32m    173\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 174\u001b[0;31m         \u001b[0muniques\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    175\u001b[0m         \u001b[0muniques\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmissing_values\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: '<' not supported between instances of 'str' and 'int'",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-16-9c978e6653fa>\u001b[0m in \u001b[0;36m<cell line: 5>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# Encoding categorical target variable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mLabelEncoder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0my_encoded\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# Fit the model using encoded target variable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py\u001b[0m in \u001b[0;36mfit_transform\u001b[0;34m(self, y)\u001b[0m\n\u001b[1;32m    113\u001b[0m         \"\"\"\n\u001b[1;32m    114\u001b[0m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcolumn_or_1d\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwarn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclasses_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_inverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    116\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m_unique\u001b[0;34m(values, return_inverse, return_counts)\u001b[0m\n\u001b[1;32m     40\u001b[0m     \"\"\"\n\u001b[1;32m     41\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 42\u001b[0;31m         return _unique_python(\n\u001b[0m\u001b[1;32m     43\u001b[0m             \u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_inverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreturn_inverse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_counts\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreturn_counts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m         )\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m_unique_python\u001b[0;34m(values, return_inverse, return_counts)\u001b[0m\n\u001b[1;32m    177\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    178\u001b[0m         \u001b[0mtypes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__qualname__\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 179\u001b[0;31m         raise TypeError(\n\u001b[0m\u001b[1;32m    180\u001b[0m             \u001b[0;34m\"Encoders require their input argument must be uniformly \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    181\u001b[0m             \u001b[0;34mf\"strings or numbers. Got {types}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: Encoders require their input argument must be uniformly strings or numbers. Got ['int', 'str']"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Convert all values in y_train and y_test to strings\n",
        "y_train = y_train.astype(str)\n",
        "y_test = y_test.astype(str)\n",
        "\n",
        "# Now use LabelEncoder to encode the categorical labels\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "\n",
        "# Encoding categorical target variable\n",
        "le = LabelEncoder()\n",
        "y_train_encoded = le.fit_transform(y_train)\n",
        "y_test_encoded = le.transform(y_test)\n",
        "\n",
        "# Proceed with model training\n",
        "rf.fit(X_train, y_train_encoded)\n",
        "\n",
        "# Prediction and Evaluation\n",
        "y_pred = rf.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test_encoded, y_pred))\n",
        "print(classification_report(y_test_encoded, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 418
        },
        "id": "uDozG2W4z6cI",
        "outputId": "9ecaf99d-aa18-45b4-a359-54770f91f5a2"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "y contains previously unseen labels: 'TRANSCARBAMYLASE'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m_encode\u001b[0;34m(values, uniques, check_unknown)\u001b[0m\n\u001b[1;32m    224\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 225\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0m_map_to_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    226\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m_map_to_integer\u001b[0;34m(values, uniques)\u001b[0m\n\u001b[1;32m    164\u001b[0m     \u001b[0mtable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_nandict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 165\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtable\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    166\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    164\u001b[0m     \u001b[0mtable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_nandict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 165\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtable\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    166\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m__missing__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    158\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnan_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 159\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    160\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyError\u001b[0m: 'TRANSCARBAMYLASE'",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-17-4f4215b026e8>\u001b[0m in \u001b[0;36m<cell line: 11>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mLabelEncoder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0my_train_encoded\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m \u001b[0my_test_encoded\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;31m# Proceed with model training\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py\u001b[0m in \u001b[0;36mtransform\u001b[0;34m(self, y)\u001b[0m\n\u001b[1;32m    135\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 137\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_encode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muniques\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclasses_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    138\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    139\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0minverse_transform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_encode.py\u001b[0m in \u001b[0;36m_encode\u001b[0;34m(values, uniques, check_unknown)\u001b[0m\n\u001b[1;32m    225\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0m_map_to_integer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0muniques\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    226\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 227\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"y contains previously unseen labels: {str(e)}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    228\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    229\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcheck_unknown\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: y contains previously unseen labels: 'TRANSCARBAMYLASE'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "import numpy as np\n",
        "\n",
        "# Combine y_train and y_test for fitting the encoder\n",
        "y_combined = pd.concat([y_train, y_test])\n",
        "\n",
        "# Fit the LabelEncoder on the combined data\n",
        "le = LabelEncoder()\n",
        "le.fit(y_combined)\n",
        "\n",
        "# Transform y_train and y_test separately\n",
        "y_train_encoded = le.transform(y_train)\n",
        "y_test_encoded = le.transform(y_test)\n",
        "\n",
        "# Now proceed with model training\n",
        "rf.fit(X_train, y_train_encoded)\n",
        "\n",
        "# Prediction and Evaluation\n",
        "y_pred = rf.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test_encoded, y_pred))\n",
        "print(classification_report(y_test_encoded, y_pred))\n"
      ],
      "metadata": {
        "id": "xJz8eVP2z9DP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Sample 10% of the data\n",
        "sampled_data = merged_data.sample(frac=0.1, random_state=42)\n",
        "\n",
        "# Proceed with the sampled data\n",
        "X = sampled_data[['sequence_length', 'residueCount_x']]\n",
        "y = sampled_data['classification']\n"
      ],
      "metadata": {
        "id": "5l8n6ZAD1MeO"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Downcast numerical columns to save memory\n",
        "merged_data['residueCount_x'] = pd.to_numeric(merged_data['residueCount_x'], downcast='integer')\n",
        "merged_data['resolution'] = pd.to_numeric(merged_data['resolution'], downcast='float')\n",
        "\n",
        "# Check memory usage\n",
        "print(merged_data.memory_usage(deep=True))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LkgAMHV_1OsI",
        "outputId": "056fc380-74be-4d01-a319-f4000a0ef20a"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Index                             128\n",
            "structureId                  28740089\n",
            "classification               33675773\n",
            "experimentalTechnique        34893444\n",
            "macromoleculeType_x          29334644\n",
            "residueCount_x                1884596\n",
            "resolution                    1884596\n",
            "structureMolecularWeight      3769192\n",
            "crystallizationMethod        31258762\n",
            "crystallizationTempK          3769192\n",
            "densityMatthews               3769192\n",
            "densityPercentSol             3769192\n",
            "pdbxDetails                  61748955\n",
            "phValue                       3769192\n",
            "publicationYear               3769192\n",
            "chainId                      27372132\n",
            "sequence                    143441710\n",
            "residueCount_y                3769192\n",
            "macromoleculeType_y          29334644\n",
            "sequence_length               3769192\n",
            "aa_composition              288228248\n",
            "dtype: int64\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pip install dask-ml dask[complete]\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "R_kuxhBu1fmJ",
        "outputId": "b17a66cd-9fb3-435e-cb70-cd281925a6ff"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting dask-ml\n",
            "  Downloading dask_ml-2024.4.4-py3-none-any.whl.metadata (5.9 kB)\n",
            "Requirement already satisfied: dask[complete] in /usr/local/lib/python3.10/dist-packages (2024.7.1)\n",
            "Collecting dask-glm>=0.2.0 (from dask-ml)\n",
            "  Downloading dask_glm-0.3.2-py2.py3-none-any.whl.metadata (1.5 kB)\n",
            "Requirement already satisfied: distributed>=2.4.0 in /usr/local/lib/python3.10/dist-packages (from dask-ml) (2024.7.1)\n",
            "Requirement already satisfied: multipledispatch>=0.4.9 in /usr/local/lib/python3.10/dist-packages (from dask-ml) (1.0.0)\n",
            "Requirement already satisfied: numba>=0.51.0 in /usr/local/lib/python3.10/dist-packages (from dask-ml) (0.60.0)\n",
            "Requirement already satisfied: numpy>=1.20.0 in /usr/local/lib/python3.10/dist-packages (from dask-ml) (1.26.4)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from dask-ml) (24.1)\n",
            "Requirement already satisfied: pandas>=0.24.2 in /usr/local/lib/python3.10/dist-packages (from dask-ml) (2.1.4)\n",
            "Requirement already satisfied: scikit-learn>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from dask-ml) (1.3.2)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from dask-ml) (1.13.1)\n",
            "Requirement already satisfied: click>=8.1 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (8.1.7)\n",
            "Requirement already satisfied: cloudpickle>=1.5.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (2.2.1)\n",
            "Requirement already satisfied: fsspec>=2021.09.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (2024.6.1)\n",
            "Requirement already satisfied: partd>=1.4.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (1.4.2)\n",
            "Requirement already satisfied: pyyaml>=5.3.1 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (6.0.1)\n",
            "Requirement already satisfied: toolz>=0.10.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (0.12.1)\n",
            "Requirement already satisfied: importlib-metadata>=4.13.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (8.2.0)\n",
            "Requirement already satisfied: pyarrow>=7.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (14.0.2)\n",
            "Requirement already satisfied: pyarrow-hotfix in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (0.6)\n",
            "Collecting lz4>=4.3.2 (from dask[complete])\n",
            "  Downloading lz4-4.3.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.7 kB)\n",
            "Collecting sparse>=0.7.0 (from dask-glm>=0.2.0->dask-ml)\n",
            "  Downloading sparse-0.15.4-py2.py3-none-any.whl.metadata (4.5 kB)\n",
            "Collecting dask-expr<1.2,>=1.1 (from dask[array,dataframe]>=2.4.0->dask-ml)\n",
            "  Downloading dask_expr-1.1.10-py3-none-any.whl.metadata (2.5 kB)\n",
            "Requirement already satisfied: jinja2>=2.10.3 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (3.1.4)\n",
            "Requirement already satisfied: locket>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (1.0.0)\n",
            "Requirement already satisfied: msgpack>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (1.0.8)\n",
            "Requirement already satisfied: psutil>=5.7.2 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (5.9.5)\n",
            "Requirement already satisfied: sortedcontainers>=2.0.5 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (2.4.0)\n",
            "Requirement already satisfied: tblib>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (3.0.0)\n",
            "Requirement already satisfied: tornado>=6.0.4 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (6.3.3)\n",
            "Requirement already satisfied: urllib3>=1.24.3 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (2.0.7)\n",
            "Requirement already satisfied: zict>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed>=2.4.0->dask-ml) (3.0.0)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=4.13.0->dask[complete]) (3.19.2)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.51.0->dask-ml) (0.43.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas>=0.24.2->dask-ml) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=0.24.2->dask-ml) (2024.1)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=0.24.2->dask-ml) (2024.1)\n",
            "Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.2.0->dask-ml) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn>=1.2.0->dask-ml) (3.5.0)\n",
            "Requirement already satisfied: bokeh>=2.4.2 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (3.4.3)\n",
            "Requirement already satisfied: contourpy>=1.2 in /usr/local/lib/python3.10/dist-packages (from bokeh>=2.4.2->dask[complete]) (1.2.1)\n",
            "Requirement already satisfied: pillow>=7.1.0 in /usr/local/lib/python3.10/dist-packages (from bokeh>=2.4.2->dask[complete]) (9.4.0)\n",
            "Requirement already satisfied: xyzservices>=2021.09.1 in /usr/local/lib/python3.10/dist-packages (from bokeh>=2.4.2->dask[complete]) (2024.6.0)\n",
            "INFO: pip is looking at multiple versions of dask-expr to determine which version is compatible with other requirements. This could take a while.\n",
            "  Downloading dask_expr-1.1.9-py3-none-any.whl.metadata (2.5 kB)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2>=2.10.3->distributed>=2.4.0->dask-ml) (2.1.5)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas>=0.24.2->dask-ml) (1.16.0)\n",
            "Downloading dask_ml-2024.4.4-py3-none-any.whl (149 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m149.8/149.8 kB\u001b[0m \u001b[31m2.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dask_glm-0.3.2-py2.py3-none-any.whl (13 kB)\n",
            "Downloading lz4-4.3.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m25.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dask_expr-1.1.9-py3-none-any.whl (241 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m241.9/241.9 kB\u001b[0m \u001b[31m13.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading sparse-0.15.4-py2.py3-none-any.whl (237 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m237.3/237.3 kB\u001b[0m \u001b[31m11.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: lz4, sparse, dask-expr, dask-glm, dask-ml\n",
            "Successfully installed dask-expr-1.1.9 dask-glm-0.3.2 dask-ml-2024.4.4 lz4-4.3.3 sparse-0.15.4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from dask_ml.model_selection import train_test_split\n",
        "from dask_ml.ensemble import RandomForestClassifier\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "ddf = dd.read_csv('/path/to/your/large_file.csv')\n",
        "\n",
        "# Split the data\n",
        "X = ddf[['sequence_length', 'residueCount_x']]\n",
        "y = ddf['classification']\n",
        "\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Initialize Dask-ML's RandomForestClassifier\n",
        "clf = RandomForestClassifier()\n",
        "\n",
        "# Fit the model (this will be done in parallel, in chunks)\n",
        "clf.fit(X_train, y_train)\n",
        "\n",
        "# Predict and evaluate the model\n",
        "y_pred = clf.predict(X_test)\n",
        "print(clf.score(X_test, y_test))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 411
        },
        "id": "0vx_C1r91mVX",
        "outputId": "ff7398be-19c5-4e43-a1e5-2272b459f4e1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ImportError",
          "evalue": "cannot import name 'RandomForestClassifier' from 'dask_ml.ensemble' (/usr/local/lib/python3.10/dist-packages/dask_ml/ensemble/__init__.py)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-12-e6031d4cb7d6>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdataframe\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mdd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdask_ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodel_selection\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mtrain_test_split\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mdask_ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mensemble\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mRandomForestClassifier\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# Load your large CSV file in chunks using Dask\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mImportError\u001b[0m: cannot import name 'RandomForestClassifier' from 'dask_ml.ensemble' (/usr/local/lib/python3.10/dist-packages/dask_ml/ensemble/__init__.py)",
            "",
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"
          ],
          "errorDetails": {
            "actions": [
              {
                "action": "open_url",
                "actionText": "Open Examples",
                "url": "/notebooks/snippets/importing_libraries.ipynb"
              }
            ]
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pip install dask[complete] scikit-learn\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Nxt_00GN1zKO",
        "outputId": "61811468-7631-41a4-be65-0b566db1e322"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.10/dist-packages (1.3.2)\n",
            "Requirement already satisfied: dask[complete] in /usr/local/lib/python3.10/dist-packages (2024.7.1)\n",
            "Requirement already satisfied: click>=8.1 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (8.1.7)\n",
            "Requirement already satisfied: cloudpickle>=1.5.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (2.2.1)\n",
            "Requirement already satisfied: fsspec>=2021.09.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (2024.6.1)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (24.1)\n",
            "Requirement already satisfied: partd>=1.4.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (1.4.2)\n",
            "Requirement already satisfied: pyyaml>=5.3.1 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (6.0.1)\n",
            "Requirement already satisfied: toolz>=0.10.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (0.12.1)\n",
            "Requirement already satisfied: importlib-metadata>=4.13.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (8.2.0)\n",
            "Requirement already satisfied: pyarrow>=7.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (14.0.2)\n",
            "Requirement already satisfied: pyarrow-hotfix in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (0.6)\n",
            "Requirement already satisfied: lz4>=4.3.2 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (4.3.3)\n",
            "Requirement already satisfied: numpy<2.0,>=1.17.3 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.26.4)\n",
            "Requirement already satisfied: scipy>=1.5.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.13.1)\n",
            "Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (3.5.0)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=4.13.0->dask[complete]) (3.19.2)\n",
            "Requirement already satisfied: locket in /usr/local/lib/python3.10/dist-packages (from partd>=1.4.0->dask[complete]) (1.0.0)\n",
            "Requirement already satisfied: pandas>=2.0 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (2.1.4)\n",
            "Requirement already satisfied: dask-expr<1.2,>=1.1 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (1.1.9)\n",
            "Requirement already satisfied: bokeh>=2.4.2 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (3.4.3)\n",
            "Requirement already satisfied: jinja2>=2.10.3 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (3.1.4)\n",
            "Requirement already satisfied: distributed==2024.7.1 in /usr/local/lib/python3.10/dist-packages (from dask[complete]) (2024.7.1)\n",
            "Requirement already satisfied: msgpack>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (1.0.8)\n",
            "Requirement already satisfied: psutil>=5.7.2 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (5.9.5)\n",
            "Requirement already satisfied: sortedcontainers>=2.0.5 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (2.4.0)\n",
            "Requirement already satisfied: tblib>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (3.0.0)\n",
            "Requirement already satisfied: tornado>=6.0.4 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (6.3.3)\n",
            "Requirement already satisfied: urllib3>=1.24.3 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (2.0.7)\n",
            "Requirement already satisfied: zict>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed==2024.7.1->dask[complete]) (3.0.0)\n",
            "Requirement already satisfied: contourpy>=1.2 in /usr/local/lib/python3.10/dist-packages (from bokeh>=2.4.2->dask[complete]) (1.2.1)\n",
            "Requirement already satisfied: pillow>=7.1.0 in /usr/local/lib/python3.10/dist-packages (from bokeh>=2.4.2->dask[complete]) (9.4.0)\n",
            "Requirement already satisfied: xyzservices>=2021.09.1 in /usr/local/lib/python3.10/dist-packages (from bokeh>=2.4.2->dask[complete]) (2024.6.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2>=2.10.3->dask[complete]) (2.1.5)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas>=2.0->dask[complete]) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=2.0->dask[complete]) (2024.1)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas>=2.0->dask[complete]) (2024.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas>=2.0->dask[complete]) (1.16.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "pdb_data_no_dups = pd.read_csv('./pdb_data_no_dups.csv')\n",
        "pdb_data_seq = pd.read_csv('./pdb_data_seq.csv')\n",
        "\n",
        "# Merge the datasets on 'structureId'\n",
        "merged_data = pd.merge(pdb_data_no_dups, pdb_data_seq, on='structureId')\n",
        "ddf = merged_data\n",
        "\n",
        "# Define your features and target\n",
        "X = ddf[['sequence_length', 'residueCount_x']]\n",
        "y = ddf['classification']\n",
        "\n",
        "# Initialize the model\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate over chunks of data\n",
        "for X_chunk, y_chunk in ddf.to_delayed().compute():\n",
        "    # Convert chunk to numpy arrays\n",
        "    X_chunk = X_chunk.to_numpy()\n",
        "    y_chunk = y_chunk.to_numpy()\n",
        "\n",
        "    # Incremental fitting of the model\n",
        "    model.partial_fit(X_chunk, y_chunk, classes=np.unique(y.compute()))\n",
        "\n",
        "# Predict on the entire test set (using Dask for scalability)\n",
        "y_pred = model.predict(X.compute())\n",
        "\n",
        "# Evaluate the model\n",
        "print(\"Accuracy:\", accuracy_score(y.compute(), y_pred))\n",
        "print(classification_report(y.compute(), y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 365
        },
        "id": "plggAj_E14az",
        "outputId": "84f95a3c-cafe-45e5-9bbe-aa6ac91f3c3f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "KeyError",
          "evalue": "\"['sequence_length'] not in index\"",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-15-66a76f993341>\u001b[0m in \u001b[0;36m<cell line: 14>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;31m# Define your features and target\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'residueCount_x'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'classification'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3897\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3898\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3899\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_indexer_strict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"columns\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3900\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3901\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6113\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6115\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_if_missing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6117\u001b[0m         \u001b[0mkeyarr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6177\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6178\u001b[0m             \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmissing_mask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6179\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{not_found} not in index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6180\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6181\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0moverload\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyError\u001b[0m: \"['sequence_length'] not in index\""
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "ddf = merged_data\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length' column\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "ddf['sequence_length'] = ddf['sequence'].apply(len, meta=('sequence', 'int64'))\n",
        "\n",
        "# Inspect the columns to ensure 'sequence_length' exists\n",
        "print(ddf.columns)\n",
        "\n",
        "# Define your features and target\n",
        "X = ddf[['sequence_length', 'residueCount_x']]\n",
        "y = ddf['classification']\n",
        "\n",
        "# Initialize the model\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate over chunks of data\n",
        "for X_chunk, y_chunk in ddf.to_delayed().compute():\n",
        "    # Convert chunk to numpy arrays\n",
        "    X_chunk = X_chunk.to_numpy()\n",
        "    y_chunk = y_chunk.to_numpy()\n",
        "\n",
        "    # Incremental fitting of the model\n",
        "    model.partial_fit(X_chunk, y_chunk, classes=np.unique(y.compute()))\n",
        "\n",
        "# Predict on the entire test set (using Dask for scalability)\n",
        "y_pred = model.predict(X.compute())\n",
        "\n",
        "# Evaluate the model\n",
        "print(\"Accuracy:\", accuracy_score(y.compute(), y_pred))\n",
        "print(classification_report(y.compute(), y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 400
        },
        "id": "0IA2aymP2Z1M",
        "outputId": "63da7d5e-84a2-4f13-d1ce-6723ad61063f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "len() takes no keyword arguments",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-16-ffc60fcbcf76>\u001b[0m in \u001b[0;36m<cell line: 9>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# Ensure 'sequence' column is of type str and create 'sequence_length' column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeta\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'int64'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;31m# Inspect the columns to ensure 'sequence_length' exists\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m   4762\u001b[0m             \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4763\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4764\u001b[0;31m         ).apply()\n\u001b[0m\u001b[1;32m   4765\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4766\u001b[0m     def _reindex_indexer(\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1207\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1208\u001b[0m         \u001b[0;31m# self.func is Callable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1209\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1210\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1211\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1287\u001b[0m         \u001b[0;31m#  Categorical (GH51645).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1288\u001b[0m         \u001b[0maction\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"ignore\"\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCategoricalDtype\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1289\u001b[0;31m         mapped = obj._map_values(\n\u001b[0m\u001b[1;32m   1290\u001b[0m             \u001b[0mmapper\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcurried\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1291\u001b[0m         )\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/base.py\u001b[0m in \u001b[0;36m_map_values\u001b[0;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m    919\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_action\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    920\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 921\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0malgorithms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_action\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    922\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    923\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/algorithms.py\u001b[0m in \u001b[0;36mmap_array\u001b[0;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m   1812\u001b[0m     \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1813\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mna_action\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1814\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1815\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1816\u001b[0m         return lib.map_infer_mask(\n",
            "\u001b[0;32mlib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mcurried\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m   1276\u001b[0m             \u001b[0;31m# _map_values does not support args/kwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1277\u001b[0m             \u001b[0;32mdef\u001b[0m \u001b[0mcurried\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1278\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1279\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1280\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: len() takes no keyword arguments"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "ddf = merged_data\n",
        "\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length' column\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "ddf['sequence_length'] = ddf['sequence'].apply(lambda x: len(x), meta=('sequence_length', 'int64'))\n",
        "\n",
        "# Inspect the columns to ensure 'sequence_length' exists\n",
        "print(ddf.columns)\n",
        "\n",
        "# Define your features and target\n",
        "X = ddf[['sequence_length', 'residueCount_x']]\n",
        "y = ddf['classification']\n",
        "\n",
        "# Initialize the m\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 400
        },
        "id": "RtxSEXnG2mHX",
        "outputId": "483d9ce7-59f7-41d8-d4f7-7ad284676b1c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "TypeError",
          "evalue": "<lambda>() got an unexpected keyword argument 'meta'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-17-615e1b6833bf>\u001b[0m in \u001b[0;36m<cell line: 10>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m# Ensure 'sequence' column is of type str and create 'sequence_length' column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeta\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'int64'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m# Inspect the columns to ensure 'sequence_length' exists\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m   4762\u001b[0m             \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4763\u001b[0m             \u001b[0mkwargs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4764\u001b[0;31m         ).apply()\n\u001b[0m\u001b[1;32m   4765\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4766\u001b[0m     def _reindex_indexer(\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1207\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1208\u001b[0m         \u001b[0;31m# self.func is Callable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1209\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1210\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1211\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0magg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1287\u001b[0m         \u001b[0;31m#  Categorical (GH51645).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1288\u001b[0m         \u001b[0maction\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"ignore\"\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCategoricalDtype\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1289\u001b[0;31m         mapped = obj._map_values(\n\u001b[0m\u001b[1;32m   1290\u001b[0m             \u001b[0mmapper\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcurried\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maction\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1291\u001b[0m         )\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/base.py\u001b[0m in \u001b[0;36m_map_values\u001b[0;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m    919\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_action\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    920\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 921\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0malgorithms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_action\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_action\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    922\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    923\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/algorithms.py\u001b[0m in \u001b[0;36mmap_array\u001b[0;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[1;32m   1812\u001b[0m     \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1813\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mna_action\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1814\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmapper\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1815\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1816\u001b[0m         return lib.map_infer_mask(\n",
            "\u001b[0;32mlib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/apply.py\u001b[0m in \u001b[0;36mcurried\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m   1276\u001b[0m             \u001b[0;31m# _map_values does not support args/kwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1277\u001b[0m             \u001b[0;32mdef\u001b[0m \u001b[0mcurried\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1278\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1279\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1280\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mTypeError\u001b[0m: <lambda>() got an unexpected keyword argument 'meta'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "ddf = merged_data\n",
        "\n",
        "# Ensure 'sequence' column is of type str\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "\n",
        "# Create 'sequence_length' column using map_partitions to apply len function\n",
        "ddf['sequence_length'] = ddf['sequence'].map_partitions(lambda df: df.apply(len), meta=('sequence_length', 'int64'))\n",
        "\n",
        "# Inspect the columns to ensure 'sequence_length' exists\n",
        "print(ddf.columns)\n",
        "\n",
        "# Define your features and target\n",
        "X = ddf[['sequence_length', 'residueCount_x']]\n",
        "y = ddf['classification']\n",
        "\n",
        "# Initialize the model\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate over chunks of data\n",
        "for X_chunk, y_chunk in zip(X.to_delayed(), y.to_delayed()):\n",
        "    X_chunk = X_chunk.compute()\n",
        "    y_chunk = y_chunk.compute()\n",
        "\n",
        "    # Incremental fitting of the model\n",
        "    model.partial_fit(X_chunk, y_chunk, classes=y.unique().compute())\n",
        "\n",
        "# Predict on the entire test set (using Dask for scalability)\n",
        "y_pred = model.predict(X.compute())\n",
        "\n",
        "# Evaluate the model\n",
        "print(\"Accuracy:\", accuracy_score(y.compute(), y_pred))\n",
        "print(classification_report(y.compute(), y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 341
        },
        "id": "_D6q21EW204o",
        "outputId": "e1b72602-7b0a-4068-c074-445164525ae4"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/dask/dataframe/__init__.py:42: FutureWarning: \n",
            "Dask dataframe query planning is disabled because dask-expr is not installed.\n",
            "\n",
            "You can install it with `pip install dask[dataframe]` or `conda install dask`.\n",
            "This will raise in a future version.\n",
            "\n",
            "  warnings.warn(msg, FutureWarning)\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'merged_data' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-1-c33dc39f1996>\u001b[0m in \u001b[0;36m<cell line: 6>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# Load your large CSV file in chunks using Dask\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mddf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmerged_data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m# Ensure 'sequence' column is of type str\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'merged_data' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "ddf = merged_data\n",
        "\n",
        "# Ensure 'sequence' column is of type str\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "\n",
        "# Create 'sequence_length' column by applying len() using map\n",
        "ddf['sequence_length'] = ddf['sequence'].map(len)\n",
        "\n",
        "# Inspect the columns to ensure 'sequence_length' exists\n",
        "print(ddf.columns)\n",
        "\n",
        "# Define your features and target\n",
        "X = ddf[['sequence_length', 'residueCount_x']]\n",
        "y = ddf['classification']\n",
        "\n",
        "# Initialize the model\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate over chunks of data\n",
        "for X_chunk, y_chunk in zip(X.to_delayed(), y.to_delayed()):\n",
        "    X_chunk = X_chunk.compute()\n",
        "    y_chunk = y_chunk.compute()\n",
        "\n",
        "    # Incremental fitting of the model\n",
        "    model.partial_fit(X_chunk, y_chunk, classes=y.unique().compute())\n",
        "\n",
        "# Predict on the entire test set (using Dask for scalability)\n",
        "y_pred = model.predict(X.compute())\n",
        "\n",
        "# Evaluate the model\n",
        "print(\"Accuracy:\", accuracy_score(y.compute(), y_pred))\n",
        "print(classification_report(y.compute(), y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 466
        },
        "id": "W6o1nI613HcK",
        "outputId": "6d7ffffa-33d4-4932-afa1-51a6ca0bcd5b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Index(['structureId', 'classification', 'experimentalTechnique',\n",
            "       'macromoleculeType_x', 'residueCount_x', 'resolution',\n",
            "       'structureMolecularWeight', 'crystallizationMethod',\n",
            "       'crystallizationTempK', 'densityMatthews', 'densityPercentSol',\n",
            "       'pdbxDetails', 'phValue', 'publicationYear', 'chainId', 'sequence',\n",
            "       'residueCount_y', 'macromoleculeType_y', 'sequence_length'],\n",
            "      dtype='object')\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "'DataFrame' object has no attribute 'to_delayed'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-20-0df3b04febed>\u001b[0m in \u001b[0;36m<cell line: 25>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;31m# Iterate over chunks of data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mX_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_chunk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_delayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_delayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m     \u001b[0mX_chunk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mX_chunk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0my_chunk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my_chunk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6202\u001b[0m         ):\n\u001b[1;32m   6203\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6204\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6206\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'to_delayed'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file in chunks using Dask\n",
        "ddf = merged_data\n",
        "\n",
        "# Ensure 'sequence' column is of type str\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "\n",
        "# Create 'sequence_length' column by applying len() using map\n",
        "ddf['sequence_length'] = ddf['sequence'].map(len)\n",
        "\n",
        "# Convert to Pandas DataFrame in chunks and train incrementally\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate over Dask DataFrame in chunks and convert to Pandas DataFrame\n",
        "for i in range(0, len(ddf), 10000):\n",
        "    X_chunk = ddf[['sequence_length', 'residueCount_x']].iloc[i:i+10000].compute()\n",
        "    y_chunk = ddf['classification'].iloc[i:i+10000].compute()\n",
        "\n",
        "    if i == 0:\n",
        "        model.partial_fit(X_chunk, y_chunk, classes=y_chunk.unique())\n",
        "    else:\n",
        "        model.partial_fit(X_chunk, y_chunk)\n",
        "\n",
        "# Predict and evaluate on the entire dataset\n",
        "X = ddf[['sequence_length', 'residueCount_x']].compute()\n",
        "y = ddf['classification'].compute()\n",
        "\n",
        "y_pred = model.predict(X)\n",
        "print(\"Accuracy:\", accuracy_score(y, y_pred))\n",
        "print(classification_report(y, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 341
        },
        "id": "jMos57A_3Tu6",
        "outputId": "f2a53b63-7219-4871-e138-a90406d5bf6d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "'DataFrame' object has no attribute 'compute'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-21-85b0fd1eb1bc>\u001b[0m in \u001b[0;36m<cell line: 18>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;31m# Iterate over Dask DataFrame in chunks and convert to Pandas DataFrame\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mddf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m     \u001b[0mX_chunk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'residueCount_x'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m10000\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m     \u001b[0my_chunk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'classification'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m10000\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6202\u001b[0m         ):\n\u001b[1;32m   6203\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6204\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6206\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'compute'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file using Dask\n",
        "ddf = merged_data\n",
        "\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "# ddf['sequence_length'] = ddf['sequence'].apply(len, meta=('sequence_length', 'int64'))\n",
        "\n",
        "ddf['sequence_length'] = ddf['sequence'].map(len)\n",
        "\n",
        "# Convert to Pandas in chunks\n",
        "chunks = ddf.to_delayed()\n",
        "\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate through each chunk\n",
        "for chunk in chunks:\n",
        "    chunk = chunk.compute()\n",
        "    X_chunk = chunk[['sequence_length', 'residueCount_x']]\n",
        "    y_chunk = chunk['classification']\n",
        "\n",
        "    if model.coef_ is None:\n",
        "        model.partial_fit(X_chunk, y_chunk, classes=y_chunk.unique())\n",
        "    else:\n",
        "        model.partial_fit(X_chunk, y_chunk)\n",
        "\n",
        "# Evaluate the model\n",
        "X_full = ddf[['sequence_length', 'residueCount_x']].compute()\n",
        "y_full = ddf['classification'].compute()\n",
        "\n",
        "y_pred = model.predict(X_full)\n",
        "print(\"Accuracy:\", accuracy_score(y_full, y_pred))\n",
        "print(classification_report(y_full, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 341
        },
        "id": "cNPqYsjh3s8Z",
        "outputId": "83eba9cd-9e14-4801-95e2-ee7d3e60dc9b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "'DataFrame' object has no attribute 'to_delayed'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-23-f47b1f917237>\u001b[0m in \u001b[0;36m<cell line: 15>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# Convert to Pandas in chunks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0mchunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_delayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSGDClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloss\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'log'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6202\u001b[0m         ):\n\u001b[1;32m   6203\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6204\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6206\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'to_delayed'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Assuming 'merged_data' is your large Pandas DataFrame\n",
        "df = merged_data\n",
        "\n",
        "# Initialize the model\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Process data in chunks\n",
        "chunksize = 10000\n",
        "for start in range(0, len(df), chunksize):\n",
        "    # Extract the chunk\n",
        "    chunk = df.iloc[start:start + chunksize]\n",
        "\n",
        "    # Prepare the features and target\n",
        "    X_chunk = chunk[['sequence_length', 'residueCount_x']]\n",
        "    y_chunk = chunk['classification']\n",
        "\n",
        "    # Fit the model incrementally\n",
        "    if start == 0:\n",
        "        model.partial_fit(X_chunk, y_chunk, classes=y_chunk.unique())\n",
        "    else:\n",
        "        model.partial_fit(X_chunk, y_chunk)\n",
        "\n",
        "# Evaluate the model on the entire dataset\n",
        "X_full = df[['sequence_length', 'residueCount_x']]\n",
        "y_full = df['classification']\n",
        "\n",
        "y_pred = model.predict(X_full)\n",
        "print(\"Accuracy:\", accuracy_score(y_full, y_pred))\n",
        "print(classification_report(y_full, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 418
        },
        "id": "7PUWoOz86ZN5",
        "outputId": "6ccf9181-af6d-4415-c7d8-27b4985dfdfb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "InvalidParameterError",
          "evalue": "The 'loss' parameter of SGDClassifier must be a str among {'squared_hinge', 'squared_error', 'modified_huber', 'log_loss', 'perceptron', 'hinge', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'}. Got 'log' instead.",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mInvalidParameterError\u001b[0m                     Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-24-fdcdf5d81186>\u001b[0m in \u001b[0;36m<cell line: 13>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     21\u001b[0m     \u001b[0;31m# Fit the model incrementally\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 23\u001b[0;31m         \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartial_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mclasses\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0my_chunk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     24\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m         \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartial_fit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_chunk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1144\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mglobal_skip_validation\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpartial_fit_and_fitted\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1145\u001b[0;31m                 \u001b[0mestimator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_params\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1147\u001b[0m             with config_context(\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36m_validate_params\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    636\u001b[0m         \u001b[0maccepted\u001b[0m \u001b[0mconstraints\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    637\u001b[0m         \"\"\"\n\u001b[0;32m--> 638\u001b[0;31m         validate_parameter_constraints(\n\u001b[0m\u001b[1;32m    639\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parameter_constraints\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    640\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_params\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_param_validation.py\u001b[0m in \u001b[0;36mvalidate_parameter_constraints\u001b[0;34m(parameter_constraints, params, caller_name)\u001b[0m\n\u001b[1;32m     94\u001b[0m                 )\n\u001b[1;32m     95\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 96\u001b[0;31m             raise InvalidParameterError(\n\u001b[0m\u001b[1;32m     97\u001b[0m                 \u001b[0;34mf\"The {param_name!r} parameter of {caller_name} must be\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m                 \u001b[0;34mf\" {constraints_str}. Got {param_val!r} instead.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mInvalidParameterError\u001b[0m: The 'loss' parameter of SGDClassifier must be a str among {'squared_hinge', 'squared_error', 'modified_huber', 'log_loss', 'perceptron', 'hinge', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'}. Got 'log' instead."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import dask.dataframe as dd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Load your large CSV file using Dask\n",
        "ddf = merged_data\n",
        "\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "ddf['sequence'] = ddf['sequence'].astype(str)\n",
        "ddf['sequence_length'] = ddf['sequence'].map(len)\n",
        "\n",
        "# Convert to Dask Delayed objects for chunk processing\n",
        "chunks = ddf.to_delayed()\n",
        "\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "\n",
        "# Iterate through each chunk\n",
        "for chunk in chunks:\n",
        "    chunk = chunk.compute()\n",
        "    X_chunk = chunk[['sequence_length', 'residueCount_x']]\n",
        "    y_chunk = chunk['classification']\n",
        "\n",
        "    if 'coef_' not in dir(model):\n",
        "        model.partial_fit(X_chunk, y_chunk, classes=y_chunk.unique())\n",
        "    else:\n",
        "        model.partial_fit(X_chunk, y_chunk)\n",
        "\n",
        "# Evaluate the model\n",
        "X_full = ddf[['sequence_length', 'residueCount_x']].compute()\n",
        "y_full = ddf['classification'].compute()\n",
        "\n",
        "y_pred = model.predict(X_full)\n",
        "print(\"Accuracy:\", accuracy_score(y_full, y_pred))\n",
        "print(classification_report(y_full, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 341
        },
        "id": "oIypyleo6eXm",
        "outputId": "05c9f54c-262a-461e-be55-364ec1cad6dd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "AttributeError",
          "evalue": "'DataFrame' object has no attribute 'to_delayed'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-27-4f3154c96dea>\u001b[0m in \u001b[0;36m<cell line: 13>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;31m# Convert to Dask Delayed objects for chunk processing\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0mchunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mddf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_delayed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSGDClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloss\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'log'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6202\u001b[0m         ):\n\u001b[1;32m   6203\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6204\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6206\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'to_delayed'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Assuming 'merged_data' is your original large Pandas DataFrame\n",
        "\n",
        "# Reduce the dataset size by sampling 10% of the data\n",
        "sampled_data = merged_data.sample(frac=0.1, random_state=42)\n",
        "\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "sampled_data['sequence'] = sampled_data['sequence'].astype(str)\n",
        "sampled_data['sequence_length'] = sampled_data['sequence'].apply(len)\n",
        "\n",
        "# Split the data into features and target\n",
        "X = sampled_data[['sequence_length', 'residueCount_x']]\n",
        "y = sampled_data['classification']\n",
        "\n",
        "# Train-test split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Initialize and train the model\n",
        "model = SGDClassifier(loss='log', random_state=42)\n",
        "model.fit(X_train, y_train)\n",
        "\n",
        "# Make predictions and evaluate the model\n",
        "y_pred = model.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
        "print(classification_report(y_test, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 418
        },
        "id": "Zl3YqLVP7Flv",
        "outputId": "86a069cf-d3b2-4964-d775-d9aa8a23a20d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "InvalidParameterError",
          "evalue": "The 'loss' parameter of SGDClassifier must be a str among {'squared_hinge', 'squared_error', 'modified_huber', 'log_loss', 'perceptron', 'hinge', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'}. Got 'log' instead.",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mInvalidParameterError\u001b[0m                     Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-28-bdda058d0e46>\u001b[0m in \u001b[0;36m<cell line: 24>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;31m# Initialize and train the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0mmodel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSGDClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloss\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'log'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrandom_state\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m42\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0;31m# Make predictions and evaluate the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1144\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mglobal_skip_validation\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpartial_fit_and_fitted\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1145\u001b[0;31m                 \u001b[0mestimator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_validate_params\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1147\u001b[0m             with config_context(\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/base.py\u001b[0m in \u001b[0;36m_validate_params\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    636\u001b[0m         \u001b[0maccepted\u001b[0m \u001b[0mconstraints\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    637\u001b[0m         \"\"\"\n\u001b[0;32m--> 638\u001b[0;31m         validate_parameter_constraints(\n\u001b[0m\u001b[1;32m    639\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parameter_constraints\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    640\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_params\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_param_validation.py\u001b[0m in \u001b[0;36mvalidate_parameter_constraints\u001b[0;34m(parameter_constraints, params, caller_name)\u001b[0m\n\u001b[1;32m     94\u001b[0m                 )\n\u001b[1;32m     95\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 96\u001b[0;31m             raise InvalidParameterError(\n\u001b[0m\u001b[1;32m     97\u001b[0m                 \u001b[0;34mf\"The {param_name!r} parameter of {caller_name} must be\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m                 \u001b[0;34mf\" {constraints_str}. Got {param_val!r} instead.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mInvalidParameterError\u001b[0m: The 'loss' parameter of SGDClassifier must be a str among {'squared_hinge', 'squared_error', 'modified_huber', 'log_loss', 'perceptron', 'hinge', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'}. Got 'log' instead."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "from sklearn.linear_model import SGDClassifier\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "\n",
        "# Assuming 'merged_data' is your original large Pandas DataFrame\n",
        "\n",
        "# Reduce the dataset size by sampling 10% of the data\n",
        "sampled_data = merged_data.sample(frac=0.1, random_state=42)\n",
        "\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "sampled_data['sequence'] = sampled_data['sequence'].astype(str)\n",
        "sampled_data['sequence_length'] = sampled_data['sequence'].apply(len)\n",
        "\n",
        "# Split the data into features and target\n",
        "X = sampled_data[['sequence_length', 'residueCount_x']]\n",
        "y = sampled_data['classification']\n",
        "\n",
        "# Train-test split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Initialize and train the model\n",
        "model = SGDClassifier(loss='log_loss', random_state=42)  # Corrected loss parameter\n",
        "model.fit(X_train, y_train)\n",
        "\n",
        "# Make predictions and evaluate the model\n",
        "y_pred = model.predict(X_test)\n",
        "print(\"Accuracy:\", accuracy_score(y_test, y_pred))\n",
        "print(classification_report(y_test, y_pred))\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1fe7xgzX7Oxw",
        "outputId": "9cccfa2d-4cae-4adf-b0b2-d6600bf09a3b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Accuracy: 0.15483391701156743\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, msg_start, len(result))\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, msg_start, len(result))\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, msg_start, len(result))\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, msg_start, len(result))\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "                                                    precision    recall  f1-score   support\n",
            "\n",
            "                     ACETYLCHOLINE BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                 ACETYLCHOLINE RECEPTOR ANTAGONIST       0.00      0.00      0.00         1\n",
            "                     ACETYLCHOLINE-BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                             ACTIN BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                             ACTIN-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                   ACYLTRANSFERASE       0.00      0.00      0.00         1\n",
            "                                  ADHESION PROTEIN       0.00      0.00      0.00         1\n",
            "                              ALKALINE PHOSPHATASE       0.00      0.00      0.00         1\n",
            "                                          ALLERGEN       0.00      0.00      0.00         2\n",
            "                                  AMIDOTRANSFERASE       0.00      0.00      0.00         1\n",
            "                                     AMINE OXIDASE       0.00      0.00      0.00         1\n",
            "                                    AMINOPEPTIDASE       0.00      0.00      0.00         2\n",
            "                                  AMINOTRANSFERASE       0.00      0.00      0.00         2\n",
            "                                     ANTI-ONCOGENE       0.00      0.00      0.00         1\n",
            "                             ANTI-ONCOGENE PROTEIN       0.00      0.00      0.00         1\n",
            "                                        ANTIBIOTIC       0.00      0.00      0.00         5\n",
            "                              ANTIBIOTIC INHIBITOR       0.00      0.00      0.00         1\n",
            "                             ANTIBIOTIC RESISTANCE       0.00      0.00      0.00         2\n",
            "                                          ANTIBODY       0.00      0.00      0.00         1\n",
            "                                        ANTIFREEZE       0.00      0.00      0.00         1\n",
            "                             ANTIMICROBIAL PEPTIDE       0.00      0.00      0.00         1\n",
            "                             ANTIMICROBIAL PROTEIN       0.00      0.00      0.00         7\n",
            "                                         ANTITOXIN       0.00      0.00      0.00         3\n",
            "                                 ANTITUMOR PROTEIN       0.00      0.00      0.00         3\n",
            "               ANTITUMOR PROTEIN/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                                 ANTIVIRAL PROTEIN       0.00      0.00      0.00         9\n",
            "                       ANTIVIRAL PROTEIN/HYDROLASE       0.00      0.00      0.00         1\n",
            "                                         APOPTOSIS       0.00      0.00      0.00        22\n",
            "                               APOPTOSIS INHIBITOR       0.00      0.00      0.00         1\n",
            "                              APOPTOSIS, HYDROLASE       0.00      0.00      0.00         1\n",
            "                      APOPTOSIS, TRANSPORT PROTEIN       0.00      0.00      0.00         1\n",
            "                     APOPTOSIS/APOPTOSIS INHIBITOR       0.00      0.00      0.00         1\n",
            "                     APOPTOSIS/APOPTOSIS REGULATOR       0.00      0.00      0.00         1\n",
            "                       APOPTOSIS/TRANSPORT PROTEIN       0.00      0.00      0.00         1\n",
            "                                 APOPTOSIS/peptide       0.00      0.00      0.00         2\n",
            "                                 ASPARTYL PROTEASE       0.00      0.00      0.00         2\n",
            "                                      ATP SYNTHASE       0.00      0.00      0.00         1\n",
            "                                  ATPASE INHIBITOR       0.00      0.00      0.00         2\n",
            "             Acetylcholine-binding protein/agonist       0.00      0.00      0.00         2\n",
            "                                    BETA-LACTAMASE       0.00      0.00      0.00         1\n",
            "                                   BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                              BIOSYNTHETIC PROTEIN       0.00      0.00      0.00        21\n",
            "                 BIOSYNTHETIC PROTEIN, RNA BINDING       0.00      0.00      0.00         2\n",
            "           BIOSYNTHETIC PROTEIN,Structural Protein       0.00      0.00      0.00         1\n",
            "                  BIOSYNTHETIC PROTEIN/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                            BIOTIN BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                            BIOTIN-BINDING PROTEIN       0.00      0.00      0.00         4\n",
            "                                    BLOOD CLOTTING       0.00      0.00      0.00        16\n",
            "                  BLOOD CLOTTING/ANTITUMOR PROTEIN       0.00      0.00      0.00         1\n",
            "                BLOOD CLOTTING/hydrolase inhibitor       0.00      0.00      0.00         1\n",
            "                                 BLOOD COAGULATION       0.00      0.00      0.00         1\n",
            "             BLOOD COAGULATION/HYDROLASE INHIBITOR       0.00      0.00      0.00         1\n",
            "                            Biotin-binding protein       0.00      0.00      0.00         3\n",
            "                           CALCIUM BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                           CALCIUM-BINDING PROTEIN       0.00      0.00      0.00         4\n",
            "                      CALCIUM/PHOSPHOLIPID-BINDING       0.00      0.00      0.00         1\n",
            "                        CARBOXYLIC ESTER HYDROLASE       0.00      0.00      0.00         1\n",
            "                                       CARDIOTOXIN       0.00      0.00      0.00         1\n",
            "                                CATALYTIC ANTIBODY       0.00      0.00      0.00         1\n",
            "                                              CCDB       0.00      0.00      0.00         1\n",
            "                                     CELL ADHESION       0.00      0.00      0.00        50\n",
            "                             CELL ADHESION PROTEIN       0.00      0.00      0.00         1\n",
            "                  CELL ADHESION, SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                 CELL ADHESION, STRUCTURAL PROTEIN       0.00      0.00      0.00         2\n",
            "                       CELL ADHESION/CELL INVASION       0.00      0.00      0.00         1\n",
            "                       CELL ADHESION/IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "               CELL ADHESION/IMMUNE SYSTEM/PEPTIDE       0.00      0.00      0.00         1\n",
            "                                        CELL CYCLE       0.00      0.00      0.00        57\n",
            "                                    CELL CYCLE/DNA       0.00      0.00      0.00         3\n",
            "                              CELL CYCLE/INHIBITOR       0.00      0.00      0.00         1\n",
            "                                CELL CYCLE/Peptide       0.00      0.00      0.00         2\n",
            "                      CELL CYCLE/SIGNALING PROTEIN       0.00      0.00      0.00         2\n",
            "                            CELL CYCLE/TRANSFERASE       0.00      0.00      0.00         3\n",
            "                                     CELL INVASION       0.00      0.00      0.00         2\n",
            "                       CELL INVASION/CELL ADHESION       0.00      0.00      0.00         1\n",
            "                           CELL INVASION/INHIBITOR       0.00      0.00      0.00         1\n",
            "                             CELL MOTILITY PROTEIN       0.00      0.00      0.00         2\n",
            "                            CELL REGULATOR PROTEIN       0.00      0.00      0.00         1\n",
            "             CELLULAR LIPOPHILIC TRANSPORT PROTEIN       0.00      0.00      0.00         1\n",
            "                                         CHAPERONE       0.00      0.00      0.00        81\n",
            "                                 CHAPERONE PROTEIN       0.00      0.00      0.00         1\n",
            "                      CHAPERONE, PROTEIN TRANSPORT       0.00      0.00      0.00         1\n",
            "                               CHAPERONE/HYDROLASE       0.00      0.00      0.00         2\n",
            "                               CHAPERONE/Inhibitor       0.00      0.00      0.00         1\n",
            "                 CHAPERONE/PEPTIDE BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                 CHAPERONE/TRANSCRIPTION INHIBITOR       0.00      0.00      0.00         1\n",
            "                                        CHAPERONIN       0.00      0.00      0.00         2\n",
            "                                         CHEMOKINE       0.00      0.00      0.00         1\n",
            "              CHEMOKINE (CHEMOATTRACTANT CYTOKINE)       0.00      0.00      0.00         2\n",
            "                                        CHEMOTAXIS       0.00      0.00      0.00         2\n",
            "                     CHOLINE BINDING PROTEIN/TOXIN       0.00      0.00      0.00         1\n",
            "                            CHOLINE-BINDING DOMAIN       0.00      0.00      0.00         1\n",
            "                            CHROMOSOME SEGREGATION       0.00      0.00      0.00         3\n",
            "                           CIRCADIAN CLOCK PROTEIN       0.00      0.00      0.00         1\n",
            "              CIRCADIAN CLOCK PROTEIN, TRANSFERASE       0.00      0.00      0.00         2\n",
            "              COMPLEX (CAPSID PROTEIN/RNA HAIRPIN)       0.00      0.00      0.00         2\n",
            "               COMPLEX (CYTOKINE RECEPTOR/PEPTIDE)       0.00      0.00      0.00         1\n",
            "             COMPLEX (DNA BINDING PROTEIN/PEPTIDE)       0.00      0.00      0.00         1\n",
            "               COMPLEX (IMMUNOGLOBULIN IGG1/IGG2A)       0.00      0.00      0.00         1\n",
            "              COMPLEX (IMMUNOGLOBULIN/AUTOANTIGEN)       0.00      0.00      0.00         1\n",
            "                COMPLEX (IMMUNOGLOBULIN/HYDROLASE)       0.00      0.00      0.00         1\n",
            "           COMPLEX (IMMUNORECEPTOR/IMMUNOGLOBULIN)       0.00      0.00      0.00         1\n",
            "                COMPLEX (INHIBITOR PROTEIN/KINASE)       0.00      0.00      0.00         1\n",
            "                          COMPLEX (KINASE/PEPTIDE)       0.00      0.00      0.00         1\n",
            "               COMPLEX (METALLOPROTEASE/INHIBITOR)       0.00      0.00      0.00         1\n",
            "                     COMPLEX (MHC CLASS I/PEPTIDE)       0.00      0.00      0.00         1\n",
            "                           COMPLEX (MHC I/PEPTIDE)       0.00      0.00      0.00         1\n",
            "                  COMPLEX (OXIDOREDUCTASE/PEPTIDE)       0.00      0.00      0.00         2\n",
            "            COMPLEX (PHOSPHOTRANSFERASE/INHIBITOR)       0.00      0.00      0.00         1\n",
            "                      COMPLEX (PROTEASE/INHIBITOR)       0.00      0.00      0.00         1\n",
            "                    COMPLEX (PROTEINASE/INHIBITOR)       0.00      0.00      0.00         1\n",
            "            COMPLEX (PROTO-ONCOGENE/EARLY PROTEIN)       0.00      0.00      0.00         1\n",
            "               COMPLEX (SERINE PROTEASE/INHIBITOR)       0.00      0.00      0.00         1\n",
            "              COMPLEX (TRANSFERASE/PHOSPHOCARRIER)       0.00      0.00      0.00         1\n",
            "           COMPLEX(BIOTIN-BINDING PROTEIN/PEPTIDE)       0.00      0.00      0.00         1\n",
            "                               CONTRACTILE PROTEIN       0.00      0.00      0.00        27\n",
            "               CONTRACTILE PROTEIN/PROTEIN BINDING       0.00      0.00      0.00         2\n",
            "            CONTRACTILE PROTEIN/STRUCTURAL PROTEIN       0.00      0.00      0.00         1\n",
            "                           CYCLIN DEPENDENT KINASE       0.00      0.00      0.00         1\n",
            "                                          CYTOKINE       0.00      0.00      0.00        23\n",
            "                        CYTOKINE RECEPTOR/CYTOKINE       0.00      0.00      0.00         1\n",
            "          CYTOKINE, HORMONE/GROWTH FACTOR receptor       0.00      0.00      0.00         1\n",
            "                               CYTOKINE/ANTAGONIST       0.00      0.00      0.00         1\n",
            "                        CYTOKINE/CYTOKINE RECEPTOR       0.00      0.00      0.00         3\n",
            "                                 CYTOLYTIC PROTEIN       0.00      0.00      0.00         2\n",
            "                                CYTOSKETAL PROTEIN       0.00      0.00      0.00         1\n",
            "                      Carbohydrate-binding protein       0.00      0.00      0.00         1\n",
            "                       Cell adhesion/Immune system       0.00      0.00      0.00         1\n",
            "                Chaperone, Peptide Binding Protein       0.00      0.00      0.00         1\n",
            "                   Chaperone/Antimicrobial protein       0.00      0.00      0.00         1\n",
            "         Contractile Protein/Actin-Binding Protein       0.00      0.00      0.00         2\n",
            "                        Cytokine/Cytokine receptor       0.00      0.00      0.00         1\n",
            "                            Cytokine/Immune System       0.00      0.00      0.00         1\n",
            "                                   DE NOVO PROTEIN       0.00      0.00      0.00        29\n",
            "            DE NOVO PROTEIN, ANTIMICROBIAL PROTEIN       0.00      0.00      0.00         1\n",
            "                                      DEHALOGENASE       0.00      0.00      0.00         2\n",
            "                                       DIOXYGENASE       0.00      0.00      0.00         2\n",
            "                                               DNA       0.00      0.00      0.00        92\n",
            "                                       DNA BINDING       0.00      0.00      0.00         1\n",
            "                               DNA BINDING PROTEIN       0.00      0.00      0.00        78\n",
            "              DNA BINDING PROTEIN, GENE REGULATION       0.00      0.00      0.00         1\n",
            "                DNA BINDING PROTEIN, HYDROLASE/DNA       0.00      0.00      0.00         2\n",
            "           DNA BINDING PROTEIN,RNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                    DNA BINDING PROTEIN/CELL CYCLE       0.00      0.00      0.00         2\n",
            "                           DNA BINDING PROTEIN/DNA       0.00      0.00      0.00        25\n",
            "                 DNA BINDING PROTEIN/HYDROLASE/DNA       0.00      0.00      0.00         1\n",
            "                        DNA BINDING PROTEIN/LIGASE       0.00      0.00      0.00         1\n",
            "               DNA BINDING PROTEIN/PROTEIN BINDING       0.00      0.00      0.00         2\n",
            "                            DNA BINDING PROTEN/DNA       0.00      0.00      0.00         1\n",
            "                    DNA BINDING REGULATORY PROTEIN       0.00      0.00      0.00         1\n",
            "                           DNA Binding Protein/DNA       0.00      0.00      0.00         3\n",
            "                               DNA EXCISION REPAIR       0.00      0.00      0.00         1\n",
            "                                   DNA INTEGRATION       0.00      0.00      0.00         1\n",
            "                       DNA LIGASE/DNA, TRANSFERASE       0.00      0.00      0.00         2\n",
            "                           DNA binding protein/DNA       0.00      0.00      0.00         4\n",
            "               DNA binding protein/protein binding       0.00      0.00      0.00         2\n",
            "                                          DNA, RNA       0.00      0.00      0.00         1\n",
            "                                       DNA-BINDING       0.00      0.00      0.00         1\n",
            "                    DNA-BINDING REGULATORY PROTEIN       0.00      0.00      0.00         1\n",
            "                                    DNA-RNA HYBRID       0.00      0.00      0.00         2\n",
            "                                           DNA/RNA       0.00      0.00      0.00         3\n",
            "                                   DNA/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                                 ELECTRON TRANSFER       0.00      0.00      0.00         1\n",
            "           ELECTRON TRANSFER (IRON-SULFUR PROTEIN)       0.00      0.00      0.00         1\n",
            "                                ELECTRON TRANSPORT       0.00      0.00      0.00        62\n",
            "                   ELECTRON TRANSPORT (CYTOCHROME)       0.00      0.00      0.00         1\n",
            "                 ELECTRON TRANSPORT (HEME PROTEIN)       0.00      0.00      0.00         1\n",
            "                    ELECTRON TRANSPORT(CYTOCHROME)       0.00      0.00      0.00         1\n",
            "                  ELECTRON TRANSPORT(HEME PROTEIN)       0.00      0.00      0.00         1\n",
            "                ELECTRON TRANSPORT, OXIDOREDUCTASE       0.00      0.00      0.00         1\n",
            "                ELECTRON TRANSPORT, PHOTOSYNTHESIS       0.00      0.00      0.00         5\n",
            "                 ELECTRON TRANSPORT,PHOTOSYNTHESIS       0.00      0.00      0.00         4\n",
            "                                       ENDOCYTOSIS       0.00      0.00      0.00         9\n",
            "                            ENDOCYTOSIS,EXOCYTOSIS       0.00      0.00      0.00         1\n",
            "                            ENDOCYTOSIS/EXOCYTOSIS       0.00      0.00      0.00         5\n",
            "                                      ENDONUCLEASE       0.00      0.00      0.00         1\n",
            "                                  ENDORIBONUCLEASE       0.00      0.00      0.00         1\n",
            "                                       ENTEROTOXIN       0.00      0.00      0.00         2\n",
            "                                         EPIMERASE       0.00      0.00      0.00         1\n",
            "                                        EXOCYTOSIS       0.00      0.00      0.00         1\n",
            "                                  EYE LENS PROTEIN       0.00      0.00      0.00         2\n",
            "                                FAMILY 10 XYLANASE       0.00      0.00      0.00         1\n",
            "                                    FIBRIL PROTEIN       0.00      0.00      0.00         1\n",
            "                                            FLAVIN       0.00      0.00      0.00         1\n",
            "                                       FLAVOENZYME       0.00      0.00      0.00         2\n",
            "                                      FLAVOPROTEIN       0.00      0.00      0.00         8\n",
            "                 FLAVOPROTEIN, FLUORESCENT PROTEIN       0.00      0.00      0.00         1\n",
            "                   FLAVOPROTEIN, SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                               FLUORESCENT PROTEIN       0.00      0.00      0.00        24\n",
            "                               FMN-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                 FOUR HELIX BUNDLE       0.00      0.00      0.00         2\n",
            "                                   GENE REGULATION       0.00      0.00      0.00        32\n",
            "                               GENE REGULATION/DNA       0.00      0.00      0.00         7\n",
            "                               GENE REGULATION/RNA       0.00      0.00      0.00         2\n",
            "                 GENE REGULATION/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                            GLYCOGEN PHOSPHORYLASE       0.00      0.00      0.00         1\n",
            "                                      GLYCOPROTEIN       0.00      0.00      0.00         5\n",
            "                                     GROWTH FACTOR       0.00      0.00      0.00         2\n",
            "              GROWTH FACTOR/GROWTH FACTOR RECEPTOR       0.00      0.00      0.00         3\n",
            " GTP-BINDING PROTEIN/GTP-BINDING PROTEIN INHIBITOR       0.00      0.00      0.00         2\n",
            "                                            GTPASE       0.00      0.00      0.00         2\n",
            "                            HALIDE BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                    HALOPEROXIDASE       0.00      0.00      0.00         1\n",
            "                                     HEMAGGLUTININ       0.00      0.00      0.00         1\n",
            "           HEME BINDING PROTEIN, TRANSPORT PROTEIN       0.00      0.00      0.00         1\n",
            "                           HEPARIN-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                  HNF-3 HOMOLOGUES       0.00      0.00      0.00         1\n",
            "                                           HORMONE       0.00      0.00      0.00        27\n",
            "                                  HORMONE RECEPTOR       0.00      0.00      0.00         4\n",
            "                HORMONE RECEPTOR/HORMONE ACTIVATOR       0.00      0.00      0.00         1\n",
            "                          HORMONE RECEPTOR/PEPTIDE       0.00      0.00      0.00         2\n",
            "                             HORMONE/GROWTH FACTOR       0.00      0.00      0.00        19\n",
            "                    HORMONE/GROWTH FACTOR RECEPTOR       0.00      0.00      0.00         2\n",
            "               HORMONE/GROWTH FACTOR/IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "                           HORMONE/PROTEIN BINDING       0.00      0.00      0.00         1\n",
            "                         HORMONE/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                                HUMANIZED ANTIBODY       0.00      0.00      0.00         1\n",
            "                                         HYDROLASE       0.00      0.00      0.00       952\n",
            "               HYDROLASE (ACTING ON CYCLIC AMIDES)       0.00      0.00      0.00         1\n",
            "                      HYDROLASE (CARBOXYLIC ESTER)       0.00      0.00      0.00         1\n",
            "                       HYDROLASE (METALLOPROTEASE)       0.00      0.00      0.00         1\n",
            "                            HYDROLASE (O-GLYCOSYL)       0.00      0.00      0.00         6\n",
            "                  HYDROLASE (PHOSPHORIC MONOESTER)       0.00      0.00      0.00         1\n",
            "                     HYDROLASE (SERINE PROTEINASE)       0.00      0.00      0.00         1\n",
            "                               HYDROLASE ACTIVATOR       0.00      0.00      0.00         2\n",
            "                                 HYDROLASE CHIMERA       0.00      0.00      0.00         1\n",
            "                              HYDROLASE INHIBITION       0.00      0.00      0.00         1\n",
            "                               HYDROLASE INHIBITOR       0.00      0.00      0.00         6\n",
            "                     HYDROLASE INHIBITOR/HYDROLASE       0.00      0.00      0.00         1\n",
            "               HYDROLASE INHIBITOR/PROTEIN BINDING       0.00      0.00      0.00         1\n",
            "                    HYDROLASE REGULATOR/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                          HYDROLASE(ACID PROTEASE)       0.00      0.00      0.00         1\n",
            "              HYDROLASE(ACTING ON ACID ANHYDRIDES)       0.00      0.00      0.00         1\n",
            "                    HYDROLASE(ASPARTYL PROTEINASE)       0.00      0.00      0.00         1\n",
            "                             HYDROLASE(O-GLYCOSYL)       0.00      0.00      0.00         3\n",
            "                 HYDROLASE(PHOSPHORIC DIESTER,RNA)       0.00      0.00      0.00         1\n",
            "                        HYDROLASE(SERINE PROTEASE)       0.00      0.00      0.00         1\n",
            "                  HYDROLASE, BLOOD CLOTTING, TOXIN       0.00      0.00      0.00         1\n",
            "                             HYDROLASE, CELL CYCLE       0.00      0.00      0.00         1\n",
            "                              HYDROLASE, ISOMERASE       0.00      0.00      0.00         1\n",
            "                  HYDROLASE, Metal binding protein       0.00      0.00      0.00         1\n",
            "                      HYDROLASE, SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                            HYDROLASE, TRANSLATION       0.00      0.00      0.00         1\n",
            "                              HYDROLASE/ANTIBIOTIC       0.00      0.00      0.00         8\n",
            "                                HYDROLASE/CYTOKINE       0.00      0.00      0.00         1\n",
            "                                     HYDROLASE/DNA       0.00      0.00      0.00        43\n",
            "                     HYDROLASE/DNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                          HYDROLASE/DNA-RNA hybrid       0.00      0.00      0.00         1\n",
            "                                 HYDROLASE/DNA/RNA       0.00      0.00      0.00         2\n",
            "                                 HYDROLASE/HORMONE       0.00      0.00      0.00         1\n",
            "                     HYDROLASE/HYDROLASE ACTIVATOR       0.00      0.00      0.00         1\n",
            "                      HYDROLASE/HYDROLASE INBHITOR       0.00      0.00      0.00         1\n",
            "                     HYDROLASE/HYDROLASE INHIBITOR       0.00      0.00      0.00       223\n",
            "                 HYDROLASE/HYDROLASE INHIBITOR/DNA       0.00      0.00      0.00         2\n",
            "                     HYDROLASE/HYDROLASE Inhibitor       0.00      0.00      0.00         2\n",
            "                      HYDROLASE/HYDROLASE RECEPTOR       0.00      0.00      0.00         1\n",
            "                     HYDROLASE/HYDROLASE SUBSTRATE       0.00      0.00      0.00         4\n",
            "                     HYDROLASE/HYDROLASE inhibitor       0.00      0.00      0.00         8\n",
            "                           HYDROLASE/IMMUNE SYSTEM       0.00      0.00      0.00        13\n",
            "                               HYDROLASE/INHIBITOR       0.00      0.00      0.00         8\n",
            "                                  HYDROLASE/LIGASE       0.00      0.00      0.00         1\n",
            "                   HYDROLASE/METAL BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                           HYDROLASE/MOTOR PROTEIN       0.00      0.00      0.00         1\n",
            "                                 HYDROLASE/PEPTIDE       0.00      0.00      0.00         2\n",
            "                         HYDROLASE/PROTEIN BINDING       0.00      0.00      0.00         7\n",
            "                             HYDROLASE/REPLICATION       0.00      0.00      0.00         1\n",
            "                                     HYDROLASE/RNA       0.00      0.00      0.00        18\n",
            "                                 HYDROLASE/RNA/DNA       0.00      0.00      0.00         2\n",
            "                      HYDROLASE/STRUCTURAL PROTEIN       0.00      0.00      0.00         3\n",
            "                       HYDROLASE/Signaling Protein       0.00      0.00      0.00         1\n",
            "                             HYDROLASE/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                         HYDROLASE/TRANSFERASE/DNA       0.00      0.00      0.00         1\n",
            "                       HYDROLASE/TRANSPORT PROTEIN       0.00      0.00      0.00         3\n",
            "                     HYDROLASE/hydrolase inhibitor       0.00      0.00      0.00         2\n",
            "                              HYPOTHETICAL PROTEIN       0.00      0.00      0.00         1\n",
            "                          Hormone receptor/peptide       0.00      0.00      0.00         1\n",
            "                    Hormone/Growth Factor Receptor       0.00      0.00      0.00         1\n",
            "            Hormone/Hormone Receptor/Immune System       0.00      0.00      0.00         1\n",
            "                               Hydrolase Inhibitor       0.00      0.00      0.00         2\n",
            "                               Hydrolase activator       0.00      0.00      0.00         1\n",
            "                               Hydrolase inhibitor       0.00      0.00      0.00         1\n",
            "                                  Hydrolase, Toxin       0.00      0.00      0.00         1\n",
            "                             Hydrolase-DNA complex       0.00      0.00      0.00         1\n",
            "                              Hydrolase/Antibiotic       0.00      0.00      0.00         3\n",
            "                                Hydrolase/Antibody       0.00      0.00      0.00         1\n",
            "                         Hydrolase/DE NOVO PROTEIN       0.00      0.00      0.00         1\n",
            "                                     Hydrolase/DNA       0.00      0.00      0.00        11\n",
            "                     Hydrolase/HYDROLASE INHIBITOR       0.00      0.00      0.00         1\n",
            "                                 Hydrolase/Hormone       0.00      0.00      0.00         1\n",
            "                     Hydrolase/Hydrolase Inhibitor       0.00      0.00      0.00         8\n",
            "                     Hydrolase/Hydrolase inhibitor       0.00      0.00      0.00         5\n",
            "                         Hydrolase/PROTEIN BINDING       0.00      0.00      0.00         1\n",
            "                                     Hydrolase/RNA       0.00      0.00      0.00         2\n",
            "                             Hydrolase/Replication       0.00      0.00      0.00         1\n",
            "                       Hydrolase/Signaling Protein       0.00      0.00      0.00         1\n",
            "                             Hydrolase/Transferase       0.00      0.00      0.00         1\n",
            "                               Hydrolase/Transport       0.00      0.00      0.00         1\n",
            "                       Hydrolase/Transport protein       0.00      0.00      0.00         1\n",
            "                              Hydrolase/antibiotic       0.00      0.00      0.00         1\n",
            "                     Hydrolase/hydrolase inhibitor       0.00      0.00      0.00         4\n",
            "                           Hydrolase/transcription       0.00      0.00      0.00         1\n",
            "                                     IMMUNE SYSTEM       0.00      0.00      0.00       318\n",
            "                  IMMUNE SYSTEM, ANTITUMOR PROTEIN       0.00      0.00      0.00         1\n",
            "              IMMUNE SYSTEM, SUGAR BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                      IMMUNE SYSTEM/ cell adhesion       0.00      0.00      0.00         1\n",
            "                            IMMUNE SYSTEM/CYTOKINE       0.00      0.00      0.00         2\n",
            "                                 IMMUNE SYSTEM/DNA       0.00      0.00      0.00         1\n",
            "                     IMMUNE SYSTEM/GENE REGULATION       0.00      0.00      0.00         1\n",
            "                           IMMUNE SYSTEM/HYDROLASE       0.00      0.00      0.00         3\n",
            "                           IMMUNE SYSTEM/INHIBITOR       0.00      0.00      0.00         4\n",
            "                      IMMUNE SYSTEM/OXIDOREDUCTASE       0.00      0.00      0.00         1\n",
            "                                 IMMUNE SYSTEM/RNA       0.00      0.00      0.00         1\n",
            "                             IMMUNE SYSTEM/RNA/DNA       0.00      0.00      0.00         1\n",
            "                   IMMUNE SYSTEM/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                               IMMUNE SYSTEM/TOXIN       0.00      0.00      0.00         1\n",
            "                         IMMUNE SYSTEM/TRANSFERASE       0.00      0.00      0.00         3\n",
            "                               IMMUNE SYSTEM/Toxin       0.00      0.00      0.00         1\n",
            "                       IMMUNE SYSTEM/VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                                    IMMUNOGLOBULIN       0.00      0.00      0.00        13\n",
            "           IMMUNOGLOBULIN(PART)SEQUESTERS ANTIGENS       0.00      0.00      0.00         1\n",
            "                    IMMUNOGLOBULIN-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                         INHIBITOR       0.00      0.00      0.00         1\n",
            "                                    INTRON-ENCODED       0.00      0.00      0.00         1\n",
            "                                      IRON STORAGE       0.00      0.00      0.00         4\n",
            "                                         ISOMERASE       0.00      0.00      0.00       121\n",
            "          ISOMERASE(INTRAMOLECULAR OXIDOREDUCTASE)       0.00      0.00      0.00         5\n",
            "                                  ISOMERASE, LYASE       0.00      0.00      0.00         1\n",
            "                  ISOMERASE, METAL-BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                                     ISOMERASE/DNA       0.00      0.00      0.00         2\n",
            "                          ISOMERASE/DNA/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                           ISOMERASE/DNA/INHIBITOR       0.00      0.00      0.00         1\n",
            "                       ISOMERASE/IMMUNOSUPPRESSANT       0.00      0.00      0.00         5\n",
            "                     ISOMERASE/ISOMERASE INHIBITOR       0.00      0.00      0.00         3\n",
            "                           Immune system/Hydrolase       0.00      0.00      0.00         2\n",
            "                             Immune system/peptide       0.00      0.00      0.00         1\n",
            "                                     Ion Transport       0.00      0.00      0.00         1\n",
            "                                     Isomerase/DNA       0.00      0.00      0.00         5\n",
            "                                     Isomerase/RNA       0.00      0.00      0.00         1\n",
            "                     Isomerase/RNA binding protein       0.00      0.00      0.00         1\n",
            "                           Isomerase/Viral protein       0.00      0.00      0.00         1\n",
            "                                            KINASE       0.00      0.00      0.00         2\n",
            "                                            LECTIN       0.00      0.00      0.00         7\n",
            "                                LECTIN(AGGLUTININ)       0.00      0.00      0.00         1\n",
            "                                    LEUCINE ZIPPER       0.00      0.00      0.00         1\n",
            "                            LIGAND BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "           LIGAND BINDING PROTEIN,MEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "                                            LIGASE       0.00      0.00      0.00       100\n",
            "                          LIGASE(AMIDE SYNTHETASE)       0.00      0.00      0.00         3\n",
            "                                LIGASE(SYNTHETASE)       0.00      0.00      0.00         1\n",
            "                                 LIGASE/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                                        LIGASE/DNA       0.00      0.00      0.00         1\n",
            "                    LIGASE/DNA-BINDING PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "                           LIGASE/LIGASE INHIBITOR       0.00      0.00      0.00         3\n",
            "                            LIGASE/PROTEIN BINDING       0.00      0.00      0.00         3\n",
            "                                        LIGASE/RNA       0.00      0.00      0.00         2\n",
            "                          LIGASE/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                         LIGASE/SIGNALLING PROTEIN       0.00      0.00      0.00         1\n",
            "                              LIGASE/TRNASCRIPTION       0.00      0.00      0.00         1\n",
            "                              LIGASE/VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                             LIPID BINDING PROTEIN       0.00      0.00      0.00        19\n",
            "                                 LIPID DEGRADATION       0.00      0.00      0.00         2\n",
            "                                   LIPID TRANSPORT       0.00      0.00      0.00         4\n",
            "                                         LIPOCALIN       0.00      0.00      0.00         1\n",
            "                               LUMINESCENT PROTEIN       0.00      0.00      0.00        10\n",
            "                                             LYASE       0.00      0.00      0.00       239\n",
            "                                  LYASE (OXO-ACID)       0.00      0.00      0.00         1\n",
            "                              LYASE(CARBON-OXYGEN)       0.00      0.00      0.00         1\n",
            "                  LYASE,OXIDOREDUCTASE/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                                         LYASE/DNA       0.00      0.00      0.00         3\n",
            "                             LYASE/LYASE INHIBITOR       0.00      0.00      0.00         9\n",
            "                             LYASE/LYASE inhibitor       0.00      0.00      0.00         1\n",
            "                       LYASE/METAL BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                 LYASE/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                                          LYSOZYME       0.00      0.00      0.00         1\n",
            "                                 Ligase, chaperone       0.00      0.00      0.00         2\n",
            "                           Ligase/Ligase Inhibitor       0.00      0.00      0.00         1\n",
            "                                Ligase/Transferase       0.00      0.00      0.00         1\n",
            "                                   Lyase Activator       0.00      0.00      0.00         1\n",
            "                             Lyase/Lyase Inhibitor       0.00      0.00      0.00         3\n",
            "                         MAGNETITE-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                  MAJOR HISTOCOMPATIBILITY COMPLEX       0.00      0.00      0.00         1\n",
            "                                  MEMBRANE PROTEIN       0.00      0.00      0.00       115\n",
            "               MEMBRANE PROTEIN, PROTEIN TRANSPORT       0.00      0.00      0.00         2\n",
            "               MEMBRANE PROTEIN, TRANSPORT PROTEIN       0.00      0.00      0.00         2\n",
            "              MEMBRANE PROTEIN/ADP-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                       MEMBRANE PROTEIN/EXOCYTOSIS       0.00      0.00      0.00         1\n",
            "                    MEMBRANE PROTEIN/IMMUNE SYSTEM       0.00      0.00      0.00         2\n",
            "                   MEMBRANE PROTEIN/OXIDOREDUCTASE       0.00      0.00      0.00         3\n",
            "                                MEMBRANE TRANSPORT       0.00      0.00      0.00         1\n",
            "                             METAL BINDING PROTEIN       0.00      0.00      0.00        62\n",
            "          METAL BINDING PROTEIN, SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                         METAL BINDING PROTEIN/DNA       0.00      0.00      0.00         7\n",
            "          METAL BINDING PROTEIN/ELECTRON TRANSPORT       0.00      0.00      0.00         1\n",
            "                   METAL BINDING PROTEIN/Hydrolase       0.00      0.00      0.00         2\n",
            "                   METAL BINDING PROTEIN/INHIBITOR       0.00      0.00      0.00         1\n",
            "                         METAL BINDING PROTEIN/RNA       0.00      0.00      0.00         1\n",
            "                                   METAL TRANSPORT       0.00      0.00      0.00        23\n",
            "                                  METALLOCHAPERONE       0.00      0.00      0.00         1\n",
            "                                    METALLOPROTEIN       0.00      0.00      0.00         1\n",
            "                                 METHYLTRANSFERASE       0.00      0.00      0.00         1\n",
            "                                     MOTOR PROTEIN       0.00      0.00      0.00        25\n",
            "                           MOTOR PROTEIN/APOPTOSIS       0.00      0.00      0.00         1\n",
            "                   MOTOR PROTEIN/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                                   MUTATOR PROTEIN       0.00      0.00      0.00         1\n",
            "                                 Membrance Protein       0.00      0.00      0.00         1\n",
            "                    Membrane protein/Immune system       0.00      0.00      0.00         1\n",
            "                 Metal Binding Protein/Transferase       0.00      0.00      0.00         1\n",
            "                   Metal Binding Protein/inhibitor       0.00      0.00      0.00         1\n",
            "                                        NEUROTOXIN       0.00      0.00      0.00         1\n",
            "                                   NUCLEAR PROTEIN       0.00      0.00      0.00        10\n",
            "                               NUCLEAR PROTEIN/DNA       0.00      0.00      0.00         3\n",
            "                               NUCLEAR PROTEIN/RNA       0.00      0.00      0.00         1\n",
            "                                  NUCLEAR RECEPTOR       0.00      0.00      0.00         1\n",
            "                                          NUCLEASE       0.00      0.00      0.00         1\n",
            "                            NUCLEIC ACID (HNA/RNA)       0.00      0.00      0.00         1\n",
            "              NUCLEIC ACID BINDING PROTEIN/DNA/RNA       0.00      0.00      0.00         1\n",
            "                        NUCLEOTIDE BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                        NUCLEOTIDE-BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                            NUCLEOTIDYLTRANSFERASE       0.00      0.00      0.00         2\n",
            "              NUCLEOTIDYLTRANSFERASE (DNA-BINDING)       0.00      0.00      0.00         1\n",
            "                                  ONCOGENE PROTEIN       0.00      0.00      0.00         2\n",
            "                                       ONCOPROTEIN       0.00      0.00      0.00         1\n",
            "                  ONCOPROTEIN, VIRUS/VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                                           OXIDASE       0.00      0.00      0.00         1\n",
            "                                    OXIDOREDUCTASE       0.09      0.37      0.14       702\n",
            "              OXIDOREDUCTASE (ALDEHYDE(D)-NAD+(A))       0.00      0.00      0.00         1\n",
            "             OXIDOREDUCTASE (CYTOCHROME(C)-OXYGEN)       0.00      0.00      0.00         2\n",
            "                   OXIDOREDUCTASE (NAD(A)-CHOH(D))       0.00      0.00      0.00         2\n",
            "                   OXIDOREDUCTASE (PQQ(A)-CHOH(D))       0.00      0.00      0.00         1\n",
            "              OXIDOREDUCTASE (SUPEROXIDE ACCEPTOR)       0.00      0.00      0.00         1\n",
            "                    OXIDOREDUCTASE(CHOH(D)-NAD(A))       0.00      0.00      0.00         1\n",
            "                            OXIDOREDUCTASE(COPPER)       0.00      0.00      0.00         1\n",
            "                      OXIDOREDUCTASE(FLAVOPROTEIN)       0.00      0.00      0.00         1\n",
            "                    OXIDOREDUCTASE(NAD(A)-CHOH(D))       0.00      0.00      0.00         1\n",
            "                      OXIDOREDUCTASE, FLAVOPROTEIN       0.00      0.00      0.00         1\n",
            "                             OXIDOREDUCTASE, TOXIN       0.00      0.00      0.00         1\n",
            "                       OXIDOREDUCTASE, TRANSFERASE       0.00      0.00      0.00         2\n",
            "                     OXIDOREDUCTASE, VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                          OXIDOREDUCTASE,HYDROLASE       0.00      0.00      0.00         1\n",
            "          OXIDOREDUCTASE/ OXIDOREDUCTASE INHIBITOR       0.00      0.00      0.00         2\n",
            "                         OXIDOREDUCTASE/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                          OXIDOREDUCTASE/CHAPERONE       0.00      0.00      0.00         1\n",
            "                 OXIDOREDUCTASE/ELECTRON TRANSPORT       0.00      0.00      0.00         7\n",
            "                          OXIDOREDUCTASE/INHIBITOR       0.00      0.00      0.00         3\n",
            "                          OXIDOREDUCTASE/Inhibitor       0.00      0.00      0.00         1\n",
            "                    OXIDOREDUCTASE/METAL transport       0.00      0.00      0.00         1\n",
            "           OXIDOREDUCTASE/OXIDOREDUCTASE INHIBITOR       0.00      0.00      0.00        35\n",
            "           OXIDOREDUCTASE/OXIDOREDUCTASE inhibitor       0.00      0.00      0.00         2\n",
            "                            OXIDOREDUCTASE/PEPTIDE       0.00      0.00      0.00         1\n",
            "                    OXIDOREDUCTASE/PROTEIN BINDING       0.00      0.00      0.00         2\n",
            "                                OXIDOREDUCTASE/RNA       0.00      0.00      0.00         1\n",
            "            OXIDOREDUCTASE/TRANSCRIPTION/INHIBITOR       0.00      0.00      0.00         1\n",
            "                        OXIDOREDUCTASE/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                                    OXO-ACID-LYASE       0.00      0.00      0.00         2\n",
            "                                    OXYGEN BINDING       0.00      0.00      0.00         3\n",
            "                                    OXYGEN CARRIER       0.00      0.00      0.00         1\n",
            "                                    OXYGEN STORAGE       0.00      0.00      0.00         2\n",
            "                  OXYGEN STORAGE, OXYGEN TRANSPORT       0.00      0.00      0.00         1\n",
            "                          OXYGEN STORAGE/TRANSPORT       0.00      0.00      0.00        18\n",
            "                                  OXYGEN TRANSPORT       0.00      0.00      0.00        20\n",
            "                   OXYGEN TRANSPORT,OXIDOREDUCTASE       0.00      0.00      0.00         1\n",
            "                OXYGEN TRANSPORT/TRANSPORT PROTEIN       0.00      0.00      0.00         1\n",
            "                Oxidoreductase, Electron transport       0.00      0.00      0.00         1\n",
            "                 Oxidoreductase/Electron transport       0.00      0.00      0.00         2\n",
            "                          Oxidoreductase/Inhibitor       0.00      0.00      0.00         1\n",
            "           Oxidoreductase/Oxidoreductase inhibitor       0.00      0.00      0.00         3\n",
            "                 Oxidoreductase/Structural Protein       0.00      0.00      0.00         2\n",
            "                  Oxygen storage, Oxygen transport       0.00      0.00      0.00         1\n",
            "                           PEPTIDE BINDING PROTEIN       0.00      0.00      0.00        16\n",
            "             PEPTIDE BINDING PROTEIN/IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "                              PEPTIDE NUCLEIC ACID       0.00      0.00      0.00         1\n",
            "                               PEROXIDE RESISTANCE       0.00      0.00      0.00         2\n",
            "                         PHEROMONE BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                 PHOSPHODIESTERASE       0.00      0.00      0.00         1\n",
            "                      PHOSPHOLIPID-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                      PHOSPHORIC DIESTER HYDROLASE       0.00      0.00      0.00         1\n",
            "                                PHOSPHOTRANSFERASE       0.00      0.00      0.00         1\n",
            "                                    PHOTOSYNTHESIS       0.00      0.00      0.00        72\n",
            "                 PHOTOSYNTHESIS/ELECTRON TRANSPORT       0.00      0.00      0.00         1\n",
            "                   PHOTOSYNTHESIS/MEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "                         PLANT LECTIN (AGGLUTININ)       0.00      0.00      0.00         1\n",
            "                                     PLANT PROTEIN       0.00      0.00      0.00        10\n",
            "                                    PLASMA PROTEIN       0.00      0.00      0.00         1\n",
            "                            PRESYNAPTIC NEUROTOXIN       0.00      0.00      0.00         2\n",
            "                                PROTEASE INHIBITOR       0.00      0.00      0.00         1\n",
            "                                   PROTEIN BINDING       0.00      0.00      0.00       105\n",
            "                         PROTEIN BINDING INHIBITOR       0.00      0.00      0.00         0\n",
            "                           PROTEIN BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                  PROTEIN BINDING, DE NOVO PROTEIN       0.00      0.00      0.00         1\n",
            "                      PROTEIN BINDING, TRANSFERASE       0.00      0.00      0.00         1\n",
            "                         PROTEIN BINDING/INHIBITOR       0.00      0.00      0.00         3\n",
            "                PROTEIN BINDING/LIGASE/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                 PROTEIN BINDING/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                       PROTEIN BINDING/TRANSFERASE       0.00      0.00      0.00         2\n",
            "                 PROTEIN BINDING/VIRAL PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "                         PROTEIN BINDING/VIRUS/DNA       0.00      0.00      0.00         1\n",
            "                              PROTEIN BIOSYNTHESIS       0.00      0.00      0.00         1\n",
            "                               PROTEIN DEGRADATION       0.00      0.00      0.00         1\n",
            "                                    PROTEIN FIBRIL       0.00      0.00      0.00        27\n",
            "                          PROTEIN FIBRIL REGULATOR       0.00      0.00      0.00         1\n",
            "                                   PROTEIN FOLDING       0.00      0.00      0.00         2\n",
            "                             PROTEIN TRANSLOCATION       0.00      0.00      0.00         1\n",
            "                                 PROTEIN TRANSPORT       0.00      0.00      0.00        56\n",
            "               PROTEIN TRANSPORT / NUCLEAR PROTEIN       0.00      0.00      0.00         1\n",
            "                      PROTEIN TRANSPORT, HYDROLASE       0.00      0.00      0.00         2\n",
            "             PROTEIN TRANSPORT/DNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                       PROTEIN TRANSPORT/HYDROLASE       0.00      0.00      0.00         1\n",
            "                          PROTEIN TRANSPORT/LIGASE       0.00      0.00      0.00         1\n",
            "                PROTEIN TRANSPORT/MEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "                  PROTEIN TRANSPORT/OXIDOREDUCTASE       0.00      0.00      0.00         1\n",
            "                 PROTEIN TURNOVER/PROTEIN TURNOVER       0.00      0.00      0.00         2\n",
            "                               PROTEIN-DNA COMPLEX       0.00      0.00      0.00         1\n",
            "                                       PROTEIN/DNA       0.00      0.00      0.00         3\n",
            "               PROTEINASE INHIBITOR (CHYMOTRYPSIN)       0.00      0.00      0.00         1\n",
            "                     PROTEINASE INHIBITOR(TRYPSIN)       0.00      0.00      0.00         1\n",
            "                                  PROTON TRANSPORT       0.00      0.00      0.00         5\n",
            "                 PROTON TRANSPORT,MEMBRANE PROTEIN       0.00      0.00      0.00         2\n",
            "             Penicillin-binding protein/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                              Peptide Nucleic acid       0.00      0.00      0.00         1\n",
            "                         Pheromone Binding Protein       0.00      0.00      0.00         2\n",
            "           Protein Transport/Lipid Binding Protein       0.00      0.00      0.00         1\n",
            "                               RAN-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                          RECEPTOR       0.00      0.00      0.00         7\n",
            "                                  RECEPTOR PROTEIN       0.00      0.00      0.00         1\n",
            "                                  RECEPTOR/PEPTIDE       0.00      0.00      0.00         1\n",
            "                                    RECEPTOR/TOXIN       0.00      0.00      0.00         2\n",
            "                                     RECOMBINATION       0.00      0.00      0.00         5\n",
            "                                RECOMBINATION, DNA       0.00      0.00      0.00         1\n",
            "                   RECOMBINATION,VIRAL PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "                                 RECOMBINATION/DNA       0.00      0.00      0.00         3\n",
            "                           RECOMBINATION/INHIBITOR       0.00      0.00      0.00         2\n",
            "                                         REDUCTASE       0.00      0.00      0.00         1\n",
            "                                       REPLICATION       0.00      0.00      0.00        15\n",
            "                         REPLICATION ACTIVATOR/DNA       0.00      0.00      0.00         1\n",
            "                      REPLICATION, TRANSFERASE/DNA       0.00      0.00      0.00         3\n",
            "                                   REPLICATION/DNA       0.00      0.00      0.00         2\n",
            "                                      RIBONUCLEASE       0.00      0.00      0.00         2\n",
            "                                 RIBOSOMAL PROTEIN       0.00      0.00      0.00         1\n",
            "                             RIBOSOMAL PROTEIN/RNA       0.00      0.00      0.00         1\n",
            "                                          RIBOSOME       0.30      0.97      0.45      1229\n",
            "                               RIBOSOME/ANTIBIOTIC       0.00      0.00      0.00       114\n",
            "                               RIBOSOME/Antibiotic       0.00      0.00      0.00         6\n",
            "                                RIBOSOME/HYDROLASE       0.00      0.00      0.00         1\n",
            "                                RIBOSOME/Hydrolase       0.00      0.00      0.00         1\n",
            "                                RIBOSOME/INHIBITOR       0.00      0.00      0.00         4\n",
            "                                   RIBOSOME/LIGASE       0.00      0.00      0.00         1\n",
            "                        RIBOSOME/PROTEIN TRANSPORT       0.00      0.00      0.00         2\n",
            "                        RIBOSOME/RIBOSOMAL PROTEIN       0.00      0.00      0.00         3\n",
            "                       RIBOSOME/RIBOSOME INHIBITOR       0.00      0.00      0.00         4\n",
            "                              RIBOSOME/TRANSLATION       0.00      0.00      0.00         4\n",
            "                               RIBOSOME/antibiotic       0.00      0.00      0.00         2\n",
            "                                               RNA       0.00      0.00      0.00        48\n",
            "                                RNA BINDING DOMAIN       0.00      0.00      0.00         1\n",
            "                               RNA BINDING PROTEIN       0.00      0.00      0.00        58\n",
            "                 RNA BINDING PROTEIN/IMMUNE SYSTEM       0.00      0.00      0.00         3\n",
            "                           RNA BINDING PROTEIN/RNA       0.00      0.00      0.00        19\n",
            "                 RNA BINDING PROTEIN/TRANSCRIPTION       0.00      0.00      0.00         1\n",
            "                                    RNA POLYMERASE       0.00      0.00      0.00         4\n",
            "                           RNA binding protein/RNA       0.00      0.00      0.00         4\n",
            "                       RNA binding protein/RNA/DNA       0.00      0.00      0.00         1\n",
            "                                    RNA/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                                       RNA/PROTEIN       0.00      0.00      0.00         2\n",
            "                           RNA/RNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                           RNA/VIRUS/VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                               Ribosome/Antibiotic       0.00      0.00      0.00        59\n",
            "                                Ribosome/Inhibitor       0.00      0.00      0.00         1\n",
            "                       Ribosome/Ribosome inhibitor       0.00      0.00      0.00         9\n",
            "                               Ribosome/antibiotic       0.00      0.00      0.00        19\n",
            "                                  SECRETION SYSTEM       0.00      0.00      0.00         1\n",
            "                              SEED STORAGE PROTEIN       0.00      0.00      0.00         1\n",
            "                                   SERINE PROTEASE       0.00      0.00      0.00         2\n",
            "                 SERINE PROTEASE/INHIBITOR COMPLEX       0.00      0.00      0.00         2\n",
            "                                     SERUM PROTEIN       0.00      0.00      0.00         1\n",
            "                               SIGNAL TRANSDUCTION       0.00      0.00      0.00         1\n",
            "                       SIGNAL TRANSDUCTION PROTEIN       0.00      0.00      0.00         3\n",
            "                                 SIGNALING PROTEIN       0.00      0.00      0.00       137\n",
            "     SIGNALING PROTEIN ACTIVATOR/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                       SIGNALING PROTEIN REGULATOR       0.00      0.00      0.00         1\n",
            "                      SIGNALING PROTEIN, HYDROLASE       0.00      0.00      0.00         1\n",
            "                  SIGNALING PROTEIN, IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "                    SIGNALING PROTEIN, TRANSFERASE       0.00      0.00      0.00         1\n",
            "                   SIGNALING PROTEIN, flavoprotein       0.00      0.00      0.00         1\n",
            "                     SIGNALING PROTEIN,TRANSFERASE       0.00      0.00      0.00         1\n",
            "          SIGNALING PROTEIN,TRANSFERASE/CELL CYCLE       0.00      0.00      0.00         1\n",
            "                      SIGNALING PROTEIN/ANTAGONIST       0.00      0.00      0.00         1\n",
            "           SIGNALING PROTEIN/ANTIMICROBIAL PROTEIN       0.00      0.00      0.00         1\n",
            "                        SIGNALING PROTEIN/Cytokine       0.00      0.00      0.00         1\n",
            "                         SIGNALING PROTEIN/HORMONE       0.00      0.00      0.00         1\n",
            "                       SIGNALING PROTEIN/INHIBITOR       0.00      0.00      0.00         1\n",
            "                          SIGNALING PROTEIN/LIGASE       0.00      0.00      0.00         2\n",
            "           SIGNALING PROTEIN/METAL BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                         SIGNALING PROTEIN/PEPTIDE       0.00      0.00      0.00         1\n",
            "                 SIGNALING PROTEIN/PROTEIN BINDING       0.00      0.00      0.00         2\n",
            "                             SIGNALING PROTEIN/RNA       0.00      0.00      0.00         2\n",
            "               SIGNALING PROTEIN/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "          SIGNALING PROTEIN/STRUCTURAL PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "                       SIGNALING PROTEIN/inhibitor       0.00      0.00      0.00         1\n",
            "                                   SM-LIKE PROTEIN       0.00      0.00      0.00         1\n",
            "                            SOLUTE-BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                                          SPLICING       0.00      0.00      0.00        28\n",
            "                                      SPLICING/RNA       0.00      0.00      0.00         6\n",
            "             STEROID BINDING PROTEIN/TRANSCRIPTION       0.00      0.00      0.00         1\n",
            "                               STRUCTURAL GENOMICS       0.00      0.00      0.00         2\n",
            "             STRUCTURAL GENOMICS, UNKNOWN FUNCTION       0.00      0.00      0.00        81\n",
            "             STRUCTURAL GENOMICS, Unknown function       0.00      0.00      0.00         1\n",
            "                                STRUCTURAL PROTEIN       0.00      0.00      0.00        80\n",
            "           STRUCTURAL PROTEIN, CONTRACTILE PROTEIN       0.00      0.00      0.00         1\n",
            "             STRUCTURAL PROTEIN, SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                 STRUCTURAL PROTEIN, VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                      STRUCTURAL PROTEIN/CHAPERONE       0.00      0.00      0.00         2\n",
            "                            STRUCTURAL PROTEIN/DNA       0.00      0.00      0.00        25\n",
            "                  STRUCTURAL PROTEIN/HYDROLASE/DNA       0.00      0.00      0.00         1\n",
            "                      STRUCTURAL PROTEIN/INHIBITOR       0.00      0.00      0.00         1\n",
            "                            STRUCTURAL PROTEIN/RNA       0.00      0.00      0.00         2\n",
            "              STRUCTURAL PROTEIN/TRANSCRIPTION/DNA       0.00      0.00      0.00         2\n",
            "            STRUCTURAL PROTEIN/VIRUS LIKE PARTICLE       0.00      0.00      0.00         1\n",
            "                                         STRUCTURE       0.00      0.00      0.00         1\n",
            "                             SUGAR BINDING PROTEIN       0.00      0.00      0.00        56\n",
            "                SUGAR BINDING, DNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                              SUPEROXIDE DISMUTASE       0.00      0.00      0.00         1\n",
            "                                          SYNTHASE       0.00      0.00      0.00         1\n",
            "                         Signaling Protein/Hormone       0.00      0.00      0.00         1\n",
            "            Signaling Protein/RHOA-BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                       Signaling protein/Hydrolase       0.00      0.00      0.00         1\n",
            "             Structural Genomics, Unknown Function       0.00      0.00      0.00        12\n",
            "             Structural Genomics, Unknown function       0.00      0.00      0.00         1\n",
            "             Structural Genomics, unknown function       0.00      0.00      0.00         1\n",
            "                            Structural PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "             Structural genomics, Unknown function       0.00      0.00      0.00         2\n",
            "             Structural genomics, unknown function       0.00      0.00      0.00         4\n",
            "              Structure Genomics, unknown function       0.00      0.00      0.00         2\n",
            "                    TATA BOX-BINDING PROTEIN (TBP)       0.00      0.00      0.00         4\n",
            "                                             TOXIN       0.00      0.00      0.00        50\n",
            "                              TOXIN RECEPTOR/TOXIN       0.00      0.00      0.00         1\n",
            "                                TOXIN, TRANSFERASE       0.00      0.00      0.00         1\n",
            "                                   TOXIN/CHAPERONE       0.00      0.00      0.00         1\n",
            "                               TOXIN/IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "                                   TOXIN/ISOMERASE       0.00      0.00      0.00         1\n",
            "                             TOXIN/TOXIN REPRESSOR       0.00      0.00      0.00         1\n",
            "                                     TRANSCRIPTION       0.00      0.00      0.00       242\n",
            "                      TRANSCRIPTION  REGULATOR/DNA       0.00      0.00      0.00         1\n",
            "                           TRANSCRIPTION ACTIVATOR       0.00      0.00      0.00         1\n",
            "                       TRANSCRIPTION ACTIVATOR/DNA       0.00      0.00      0.00         3\n",
            "                              TRANSCRIPTION FACTOR       0.00      0.00      0.00         1\n",
            "                          TRANSCRIPTION REGULATION       0.00      0.00      0.00         4\n",
            "                           TRANSCRIPTION REGULATOR       0.00      0.00      0.00        12\n",
            "    TRANSCRIPTION REGULATOR, METAL BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                       TRANSCRIPTION REGULATOR/DNA       0.00      0.00      0.00         3\n",
            "                         TRANSCRIPTION TERMINATION       0.00      0.00      0.00         1\n",
            "                           TRANSCRIPTION regulator       0.00      0.00      0.00         1\n",
            "                TRANSCRIPTION, DNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                    TRANSCRIPTION, REPLICATION/DNA       0.00      0.00      0.00         1\n",
            "            TRANSCRIPTION, RNA BINDING PROTEIN/RNA       0.00      0.00      0.00         1\n",
            "                        TRANSCRIPTION, TRANSFERASE       0.00      0.00      0.00         1\n",
            "                    TRANSCRIPTION, TRANSFERASE/DNA       0.00      0.00      0.00         4\n",
            "         TRANSCRIPTION, TRANSFERASE/DNA-RNA HYBRID       0.00      0.00      0.00         1\n",
            "                TRANSCRIPTION, TRANSFERASE/DNA/RNA       0.00      0.00      0.00         3\n",
            "                  TRANSCRIPTION, viral protein/RNA       0.00      0.00      0.00         1\n",
            "                           TRANSCRIPTION,HYDROLASE       0.00      0.00      0.00         1\n",
            "                         TRANSCRIPTION,TRANSFERASE       0.00      0.00      0.00         1\n",
            "          TRANSCRIPTION,TRANSFERASE/DNA/RNA HYBRID       0.00      0.00      0.00         1\n",
            "                                 TRANSCRIPTION/DNA       0.00      0.00      0.00        59\n",
            "                      TRANSCRIPTION/DNA-RNA HYBRID       0.00      0.00      0.00         1\n",
            "                             TRANSCRIPTION/DNA/RNA       0.00      0.00      0.00         1\n",
            "                             TRANSCRIPTION/PEPTIDE       0.00      0.00      0.00         1\n",
            "                             TRANSCRIPTION/Peptide       0.00      0.00      0.00         1\n",
            "                     TRANSCRIPTION/Protein binding       0.00      0.00      0.00         1\n",
            "                         TRANSCRIPTION/REPLICATION       0.00      0.00      0.00         1\n",
            "                                 TRANSCRIPTION/RNA       0.00      0.00      0.00         3\n",
            "                             TRANSCRIPTION/RNA/DNA       0.00      0.00      0.00         5\n",
            "                               TRANSCRIPTION/TOXIN       0.00      0.00      0.00         1\n",
            "             TRANSCRIPTION/TRANSCRIPTION INHIBITOR       0.00      0.00      0.00         2\n",
            "                         TRANSCRIPTION/TRANSFERASE       0.00      0.00      0.00         2\n",
            "                     TRANSCRIPTION/TRANSFERASE/DNA       0.00      0.00      0.00         1\n",
            "                                       TRANSFERASE       0.00      0.00      0.00       729\n",
            "               TRANSFERASE (CARBAMOYL-P,ASPARTATE)       0.00      0.00      0.00         3\n",
            "                     TRANSFERASE(AMINOTRANSFERASE)       0.00      0.00      0.00         1\n",
            "                  TRANSFERASE, DNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                            TRANSFERASE, HYDROLASE       0.00      0.00      0.00         2\n",
            "                               TRANSFERASE, LIGASE       0.00      0.00      0.00         1\n",
            "                                TRANSFERASE, LYASE       0.00      0.00      0.00         1\n",
            "                            TRANSFERASE, LYASE/DNA       0.00      0.00      0.00         2\n",
            "                       TRANSFERASE, OXIDOREDUCTASE       0.00      0.00      0.00         1\n",
            "                             TRANSFERASE,HYDROLASE       0.00      0.00      0.00         1\n",
            "                            TRANSFERASE/ANTIBIOTIC       0.00      0.00      0.00         4\n",
            "                            TRANSFERASE/CELL CYCLE       0.00      0.00      0.00         2\n",
            "                                   TRANSFERASE/DNA       0.00      0.00      0.00        66\n",
            "                           TRANSFERASE/DNA COMPLEX       0.00      0.00      0.00         1\n",
            "                       TRANSFERASE/DNA, LIGASE/DNA       0.00      0.00      0.00         2\n",
            "                               TRANSFERASE/DNA/RNA       0.00      0.00      0.00         4\n",
            "             TRANSFERASE/DNA/TRANSFERASE INHIBITOR       0.00      0.00      0.00         1\n",
            "                             TRANSFERASE/HYDROLASE       0.00      0.00      0.00         1\n",
            "                         TRANSFERASE/IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "                     TRANSFERASE/IMMUNE SYSTEM/DNA       0.00      0.00      0.00         2\n",
            "                             TRANSFERASE/INHIBITOR       0.00      0.00      0.00         6\n",
            "                                TRANSFERASE/LIGASE       0.00      0.00      0.00         1\n",
            "                 TRANSFERASE/LIPID BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                       TRANSFERASE/NUCLEAR PROTEIN       0.00      0.00      0.00         1\n",
            "                        TRANSFERASE/OXIDOREDUCTASE       0.00      0.00      0.00         1\n",
            "                  TRANSFERASE/OXIDOREDUCTASE/LYASE       0.00      0.00      0.00         1\n",
            "                               TRANSFERASE/PEPTIDE       0.00      0.00      0.00         3\n",
            "                       TRANSFERASE/PROTEIN BINDING       0.00      0.00      0.00         2\n",
            "                     TRANSFERASE/RIBOSOMAL PROTEIN       0.00      0.00      0.00         2\n",
            "                                   TRANSFERASE/RNA       0.00      0.00      0.00        14\n",
            "                               TRANSFERASE/RNA/DNA       0.00      0.00      0.00         2\n",
            "                     TRANSFERASE/SIGNALING PROTEIN       0.00      0.00      0.00         2\n",
            "                    TRANSFERASE/STRUCTURAL PROTEIN       0.00      0.00      0.00         1\n",
            "                         TRANSFERASE/TRANSCRIPTION       0.00      0.00      0.00         6\n",
            "                           TRANSFERASE/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                 TRANSFERASE/TRANSFERASE ACTIVATOR       0.00      0.00      0.00         2\n",
            "                 TRANSFERASE/TRANSFERASE INHIBITOR       0.00      0.00      0.00        59\n",
            "                 TRANSFERASE/TRANSFERASE Inhibitor       0.00      0.00      0.00         1\n",
            "                  TRANSFERASE/TRANSFERASE RECEPTOR       0.00      0.00      0.00         1\n",
            "                 TRANSFERASE/TRANSFERASE REGULATOR       0.00      0.00      0.00         2\n",
            "                 TRANSFERASE/TRANSFERASE SUBSTRATE       0.00      0.00      0.00         1\n",
            "                     TRANSFERASE/TRANSPORT PROTEIN       0.00      0.00      0.00         1\n",
            "                                       TRANSLATION       0.00      0.00      0.00        45\n",
            "                  TRANSLATION, RNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                                TRANSLATION, TOXIN       0.00      0.00      0.00         1\n",
            "              TRANSLATION/BIOSYNTHETIC PROTEIN/RNA       0.00      0.00      0.00         1\n",
            "                        TRANSLATION/OXIDOREDUCTASE       0.00      0.00      0.00         2\n",
            "                                   TRANSLATION/RNA       0.00      0.00      0.00         5\n",
            "                           TRANSLATION/TRANSFERASE       0.00      0.00      0.00         1\n",
            "                            TRANSLATION/antibiotic       0.00      0.00      0.00         1\n",
            "                             TRANSMEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "                                         TRANSPORT       0.00      0.00      0.00         5\n",
            "                                 TRANSPORT PROTEIN       0.00      0.00      0.00       182\n",
            "              TRANSPORT PROTEIN / MEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "               TRANSPORT PROTEIN, MEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "           TRANSPORT PROTEIN,SIGNALING PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "              TRANSPORT PROTEIN,STRUCTURAL PROTEIN       0.00      0.00      0.00         1\n",
            "                       TRANSPORT PROTEIN/APOPTOSIS       0.00      0.00      0.00         1\n",
            "             TRANSPORT PROTEIN/DNA binding Protein       0.00      0.00      0.00         1\n",
            "                       TRANSPORT PROTEIN/HYDROLASE       0.00      0.00      0.00         2\n",
            "                   TRANSPORT PROTEIN/IMMUNE SYSTEM       0.00      0.00      0.00         2\n",
            "                       TRANSPORT PROTEIN/INHIBITOR       0.00      0.00      0.00         3\n",
            "                       TRANSPORT PROTEIN/ISOMERASE       0.00      0.00      0.00         8\n",
            "                 TRANSPORT PROTEIN/NUCLEAR PROTEIN       0.00      0.00      0.00         1\n",
            "                 TRANSPORT PROTEIN/PROTEIN BINDING       0.00      0.00      0.00         2\n",
            "                        TRANSPORT PROTEIN/Receptor       0.00      0.00      0.00         1\n",
            "                TRANSPORT PROTEIN/Receptor/Agonist       0.00      0.00      0.00         1\n",
            "               TRANSPORT PROTEIN/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                       TRANSPORT PROTEIN/inhibitor       0.00      0.00      0.00         1\n",
            "                                   TRNA SYNTHETASE       0.00      0.00      0.00         1\n",
            "                                   Thansferase/DNA       0.00      0.00      0.00         1\n",
            "                               Toxin/CELL Adhesion       0.00      0.00      0.00         2\n",
            "                                         Toxin/RNA       0.00      0.00      0.00         1\n",
            "                                Tranferase/DNA/RNA       0.00      0.00      0.00         1\n",
            "                           Transcription regulator       0.00      0.00      0.00         2\n",
            "                Transcription, TRANSFERASE/DNA/RNA       0.00      0.00      0.00         1\n",
            "                    Transcription, protein binding       0.00      0.00      0.00         1\n",
            "                             Transcription/Agonist       0.00      0.00      0.00         1\n",
            "                                 Transcription/DNA       0.00      0.00      0.00         6\n",
            "                    Transcription/MEMBRANE PROTEIN       0.00      0.00      0.00         1\n",
            "                         Transcription/Transferase       0.00      0.00      0.00         2\n",
            "                             Transferase Inhibitor       0.00      0.00      0.00         1\n",
            "                             Transferase regulator       0.00      0.00      0.00         1\n",
            "                            Transferase, Hydrolase       0.00      0.00      0.00         1\n",
            "                             Transferase,Lyase/DNA       0.00      0.00      0.00         2\n",
            "                       Transferase,replication/DNA       0.00      0.00      0.00         1\n",
            "                            Transferase/Cell Cycle       0.00      0.00      0.00         1\n",
            "                                   Transferase/DNA       0.00      0.00      0.00        12\n",
            "                   Transferase/DNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                             Transferase/Inhibitor       0.00      0.00      0.00         2\n",
            "                                   Transferase/RNA       0.00      0.00      0.00         6\n",
            "                     Transferase/Signaling Protein       0.00      0.00      0.00         1\n",
            "                 Transferase/Transferase Inhibitor       0.00      0.00      0.00        12\n",
            "                 Transferase/Transferase inhibitor       0.00      0.00      0.00         2\n",
            "                                   Transferase/rna       0.00      0.00      0.00         2\n",
            "                 Transferase/transferase Inhibitor       0.00      0.00      0.00         1\n",
            "                 Transferase/transferase inhibitor       0.00      0.00      0.00         1\n",
            "                      Transferase/unknown function       0.00      0.00      0.00         1\n",
            "                   Transport Protein/immune System       0.00      0.00      0.00         1\n",
            "                                         UBIQUITIN       0.00      0.00      0.00         1\n",
            "                                  UNKNOWN FUNCTION       0.00      0.00      0.00        61\n",
            "                                     VIRAL PEPTIDE       0.00      0.00      0.00         1\n",
            "                                     VIRAL PROTEIN       0.00      0.00      0.00       186\n",
            "                    VIRAL PROTEIN, PROTEIN BINDING       0.00      0.00      0.00         1\n",
            "                      VIRAL PROTEIN, RECOMBINATION       0.00      0.00      0.00         1\n",
            "                VIRAL PROTEIN, RNA BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                      VIRAL PROTEIN, TRANSCRIPTION       0.00      0.00      0.00         1\n",
            "                    VIRAL PROTEIN, TRANSFERASE/RNA       0.00      0.00      0.00         1\n",
            "                       VIRAL PROTEIN/CELL ADHESION       0.00      0.00      0.00         2\n",
            "                                 VIRAL PROTEIN/DNA       0.00      0.00      0.00         1\n",
            "                       VIRAL PROTEIN/IMMUNE SYSTEM       0.00      0.00      0.00        27\n",
            "             VIRAL PROTEIN/IMMUNE SYSTEM/INHIBITOR       0.00      0.00      0.00         2\n",
            "                           VIRAL PROTEIN/INHIBITOR       0.00      0.00      0.00         1\n",
            "                       VIRAL PROTEIN/Immune system       0.00      0.00      0.00         1\n",
            "                             VIRAL PROTEIN/PEPTIDE       0.00      0.00      0.00         1\n",
            "                     VIRAL PROTEIN/PROTEIN BINDING       0.00      0.00      0.00         1\n",
            "                   VIRAL PROTEIN/PROTEIN TRANSPORT       0.00      0.00      0.00         1\n",
            "                                 VIRAL PROTEIN/RNA       0.00      0.00      0.00         5\n",
            "                 VIRAL PROTEIN/RNA BINDING PROTEIN       0.00      0.00      0.00         1\n",
            "                   VIRAL PROTEIN/SIGNALING PROTEIN       0.00      0.00      0.00         1\n",
            "                       VIRAL PROTEIN/VIRAL PROTEIN       0.00      0.00      0.00         1\n",
            "                                             VIRUS       0.00      0.00      0.00       200\n",
            "                               VIRUS LIKE PARTICLE       0.00      0.00      0.00        17\n",
            "                                         VIRUS/DNA       0.00      0.00      0.00         2\n",
            "                               VIRUS/IMMUNE SYSTEM       0.00      0.00      0.00         1\n",
            "                                    VIRUS/RECEPTOR       0.00      0.00      0.00         1\n",
            "                                         VIRUS/RNA       0.00      0.00      0.00        23\n",
            "                               VIRUS/VIRAL PROTEIN       0.00      0.00      0.00         2\n",
            "                    VIRUS/VIRAL PROTEIN/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "           VIRUS/VIRAL PROTEIN/RNA BINDING PROTEIN       0.00      0.00      0.00         2\n",
            "                        Viral Protein/Immue System       0.00      0.00      0.00         1\n",
            "                       Viral Protein/Immune System       0.00      0.00      0.00         1\n",
            "                   Viral Protein/RNA/Immune System       0.00      0.00      0.00         1\n",
            "                       Viral protein/Immune system       0.00      0.00      0.00         3\n",
            "                             Viral protein/PEPTIDE       0.00      0.00      0.00         1\n",
            "              Viral protein/receptor/Immune system       0.00      0.00      0.00         1\n",
            "                              Virus, hydrolase/RNA       0.00      0.00      0.00         1\n",
            "                               Virus/Immune system       0.00      0.00      0.00         1\n",
            "                                    Virus/Receptor       0.00      0.00      0.00         2\n",
            "                   antibiotic/antibiotic inhibitor       0.00      0.00      0.00         1\n",
            "                            biotin binding protein       0.00      0.00      0.00         1\n",
            "                           calcium binding protein       0.00      0.00      0.00         1\n",
            "                       cell adhesion/immune system       0.00      0.00      0.00         1\n",
            "                               chaperone/Inhibitor       0.00      0.00      0.00         1\n",
            "                           dna binding protein/dna       0.00      0.00      0.00         4\n",
            "                        hydrolase inhibitor, toxin       0.00      0.00      0.00         1\n",
            "                               hydrolase regulator       0.00      0.00      0.00         1\n",
            "                                     hydrolase/DNA       0.00      0.00      0.00         3\n",
            "                                     hydrolase/RNA       0.00      0.00      0.00         1\n",
            "                              hydrolase/antibiotic       0.00      0.00      0.00         2\n",
            "                                     hydrolase/dna       0.00      0.00      0.00         4\n",
            "                     hydrolase/hydrolase inhibitor       0.00      0.00      0.00        16\n",
            "                                     hydrolase/rna       0.00      0.00      0.00         1\n",
            "                                 immune system/RNA       0.00      0.00      0.00         1\n",
            "                                 ligase, lyase/DNA       0.00      0.00      0.00         1\n",
            "                           ligase/ligase inhibitor       0.00      0.00      0.00         3\n",
            "                       lyase/metal binding protein       0.00      0.00      0.00         1\n",
            "                       membrane protein, hydrolase       0.00      0.00      0.00         4\n",
            "                          membrane protein/agonist       0.00      0.00      0.00         1\n",
            "                            nickel-binding protein       0.00      0.00      0.00         1\n",
            "                oxidoreductase, electron transport       0.00      0.00      0.00         2\n",
            "                       oxidoreductase, transferase       0.00      0.00      0.00         1\n",
            "           oxidoreductase/oxidoreductase INHIBITOR       0.00      0.00      0.00         1\n",
            "           oxidoreductase/oxidoreductase inhibitor       0.00      0.00      0.00         2\n",
            "                     peptidoglycan binding protein       0.00      0.00      0.00         1\n",
            "                                protease inhibitor       0.00      0.00      0.00         1\n",
            "                    protein binding, immune system       0.00      0.00      0.00         1\n",
            "                         protein binding/hydrolase       0.00      0.00      0.00         1\n",
            "                     protein binding/immune system       0.00      0.00      0.00         1\n",
            "                           protein binding/peptide       0.00      0.00      0.00         1\n",
            "                               protein/RNA complex       0.00      0.00      0.00         1\n",
            "                                 recombination/DNA       0.00      0.00      0.00         2\n",
            "                      replication, transferase/dna       0.00      0.00      0.00         1\n",
            "                               ribosome/antibiotic       0.00      0.00      0.00        31\n",
            "                           rna binding protein/rna       0.00      0.00      0.00         2\n",
            "                         signaling Protein/Peptide       0.00      0.00      0.00         1\n",
            "signaling protein, immune system/signaling protein       0.00      0.00      0.00         1\n",
            "             signaling protein/inhibitor/activator       0.00      0.00      0.00         1\n",
            "                                  splicing/DNA/RNA       0.00      0.00      0.00         3\n",
            "             structural genomics, unknown function       0.00      0.00      0.00        14\n",
            "                                   toxin/inhibitor       0.00      0.00      0.00         1\n",
            "                        tranport protein/inhibitor       0.00      0.00      0.00         1\n",
            "               transcription regulation/ANTIBIOTIC       0.00      0.00      0.00         1\n",
            "                           transcription regulator       0.00      0.00      0.00         3\n",
            "                          transcription repression       0.00      0.00      0.00         1\n",
            "                           transcription repressor       0.00      0.00      0.00         1\n",
            "                                 transcription/DNA       0.00      0.00      0.00         3\n",
            "                                 transcription/RNA       0.00      0.00      0.00         1\n",
            "                          transcription/antibiotic       0.00      0.00      0.00         2\n",
            "                                 transcription/dna       0.00      0.00      0.00         1\n",
            "                            transferase, lyase/dna       0.00      0.00      0.00         2\n",
            "                                   transferase/DNA       0.00      0.00      0.00         3\n",
            "                                   transferase/RNA       0.00      0.00      0.00         1\n",
            "                                   transferase/dna       0.00      0.00      0.00         2\n",
            "                                   transferase/rna       0.00      0.00      0.00         2\n",
            "                 transferase/transferase inhibitor       0.00      0.00      0.00         8\n",
            "                       transport protein/inhibitor       0.00      0.00      0.00         1\n",
            "                       viral protein/immune system       0.00      0.00      0.00         2\n",
            "             viral protein/transcription inhibitor       0.00      0.00      0.00         1\n",
            "                                         virus/RNA       0.00      0.00      0.00         1\n",
            "\n",
            "                                          accuracy                           0.15      9423\n",
            "                                         macro avg       0.00      0.00      0.00      9423\n",
            "                                      weighted avg       0.05      0.15      0.07      9423\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from sklearn.svm import SVC\n",
        "\n",
        "sampled_data = merged_data.sample(frac=0.1, random_state=42)\n",
        "sampled_data['sequence'] = sampled_data['sequence'].astype(str)\n",
        "sampled_data['sequence_length'] = sampled_data['sequence'].apply(len)\n",
        "X = sampled_data[['sequence_length', 'residueCount_x']]\n",
        "y = sampled_data['classification']\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "svm_model = SVC(kernel='linear', random_state=42)\n",
        "svm_model.fit(X_train, y_train)\n",
        "y_pred_svm = svm_model.predict(X_test)\n",
        "print(\"SVM Accuracy:\", accuracy_score(y_test, y_pred_svm))\n",
        "print(classification_report(y_test, y_pred_svm))\n"
      ],
      "metadata": {
        "id": "x4rGK_lyhPee"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "sEDl4ufTpLkj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "xHIA33h8cNNZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "sQVUmCOXcNS6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pip install biopython\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "R6MjTXuGcNW8",
        "outputId": "5989fc4a-bb06-4618-9565-422a0aa95dc8"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.26.4)\n",
            "Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m22.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.84\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Blast import NCBIWWW, NCBIXML\n",
        "import pandas as pd\n",
        "\n",
        "# Load the data\n",
        "pdb_data_seq = pd.read_csv('pdb_data_seq.csv')\n",
        "\n",
        "# Extract protein sequences\n",
        "protein_sequences = pdb_data_seq[pdb_data_seq['macromoleculeType'] == 'Protein']\n",
        "\n",
        "# Function to run BLASTP\n",
        "def run_blastp(sequence, sequence_id):\n",
        "    result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", sequence)\n",
        "    with open(f\"blast_results_{sequence_id}.xml\", \"w\") as out_handle:\n",
        "        out_handle.write(result_handle.read())\n",
        "    result_handle.close()\n",
        "\n",
        "# Run BLASTP for each protein sequence\n",
        "for index, row in protein_sequences.iterrows():\n",
        "    run_blastp(row['sequence'], row['structureId'])\n",
        "\n",
        "print(\"BLASTP search completed for protein sequences.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 472
        },
        "id": "IEANcTUecSZ-",
        "outputId": "df2853bc-9226-4458-b618-7801bb9f8424"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-2-98efb747bacd>\u001b[0m in \u001b[0;36m<cell line: 18>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;31m# Run BLASTP for each protein sequence\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprotein_sequences\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m     \u001b[0mrun_blastp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'structureId'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"BLASTP search completed for protein sequences.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-2-98efb747bacd>\u001b[0m in \u001b[0;36mrun_blastp\u001b[0;34m(sequence, sequence_id)\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;31m# Function to run BLASTP\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mrun_blastp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m     \u001b[0mresult_handle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNCBIWWW\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqblast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"blastp\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"nr\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"blast_results_{sequence_id}.xml\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"w\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mout_handle\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m         \u001b[0mout_handle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult_handle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/Bio/Blast/NCBIWWW.py\u001b[0m in \u001b[0;36mqblast\u001b[0;34m(program, database, sequence, url_base, auto_format, composition_based_statistics, db_genetic_code, endpoints, entrez_query, expect, filter, gapcosts, genetic_code, hitlist_size, i_thresh, layout, lcase_mask, matrix_name, nucl_penalty, nucl_reward, other_advanced, perc_ident, phi_pattern, query_file, query_believe_defline, query_from, query_to, searchsp_eff, service, threshold, ungapped_alignment, word_size, short_query, alignments, alignment_view, descriptions, entrez_links_new_window, expect_low, expect_high, format_entrez_query, format_object, format_type, ncbi_gi, results_file, show_overview, megablast, template_type, template_length, username, password)\u001b[0m\n\u001b[1;32m    268\u001b[0m         \u001b[0mwait\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mqblast\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprevious\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mdelay\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mcurrent\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    269\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mwait\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 270\u001b[0;31m             \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    271\u001b[0m             \u001b[0mqblast\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprevious\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcurrent\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    272\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Blast import NCBIWWW, NCBIXML\n",
        "import pandas as pd\n",
        "\n",
        "# Load the data\n",
        "pdb_data_seq = pd.read_csv('pdb_data_seq.csv')\n",
        "\n",
        "# Extract protein sequences\n",
        "protein_sequences = pdb_data_seq[pdb_data_seq['macromoleculeType'] == 'Protein'].head(10)\n",
        "\n",
        "# Function to run BLASTP\n",
        "def run_blastp(sequence, sequence_id):\n",
        "    result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", sequence)\n",
        "    with open(f\"blast_results_{sequence_id}.xml\", \"w\") as out_handle:\n",
        "        out_handle.write(result_handle.read())\n",
        "    result_handle.close()\n",
        "\n",
        "# Run BLASTP for each of the first 10 protein sequences\n",
        "for index, row in protein_sequences.iterrows():\n",
        "    print(f\"Running BLASTP for sequence ID: {row['structureId']}\")\n",
        "    run_blastp(row['sequence'], row['structureId'])\n",
        "\n",
        "print(\"BLASTP search completed for the first 10 protein sequences.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "I1Ez9kdWdcTZ",
        "outputId": "576962db-bd53-4c47-d14e-8851314ff0a9"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Running BLASTP for sequence ID: 101M\n",
            "Running BLASTP for sequence ID: 102L\n",
            "Running BLASTP for sequence ID: 102M\n",
            "Running BLASTP for sequence ID: 103L\n",
            "Running BLASTP for sequence ID: 103M\n",
            "Running BLASTP for sequence ID: 104L\n",
            "Running BLASTP for sequence ID: 104L\n",
            "Running BLASTP for sequence ID: 104M\n",
            "Running BLASTP for sequence ID: 105M\n",
            "Running BLASTP for sequence ID: 106M\n",
            "BLASTP search completed for the first 10 protein sequences.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update\n",
        "!apt-get install -y ncbi-blast+\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5vqqiGjmlAq9",
        "outputId": "b876e855-d386-4c32-adda-768cd58c9f91"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rHit:1 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Connected to cloud.r-project.org (108.139.15.91)] [C\r                                                                                                    \rGet:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]\n",
            "\r                                                                                                    \rGet:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]\n",
            "\r0% [2 InRelease 21.4 kB/128 kB 17%] [3 InRelease 46.0 kB/129 kB 36%] [Connected to cloud.r-project.o\r                                                                                                    \rGet:4 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]\n",
            "\r0% [2 InRelease 63.4 kB/128 kB 50%] [3 InRelease 93.8 kB/129 kB 73%] [4 InRelease 3,626 B/3,626 B 10\r0% [2 InRelease 63.4 kB/128 kB 50%] [3 InRelease 94.2 kB/129 kB 73%] [Waiting for headers] [Waiting \r0% [2 InRelease 79.3 kB/128 kB 62%] [Waiting for headers] [Waiting for headers] [Connecting to ppa.l\r0% [Waiting for headers] [Waiting for headers] [Waiting for headers] [Connecting to ppa.launchpadcon\r                                                                                                    \rGet:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Connecting to ppa.launchpadcontent.net (185.125.190.\r                                                                                                    \rHit:6 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "\r                                                                                                    \r0% [Waiting for headers] [Connecting to ppa.launchpadcontent.net (185.125.190.80)]\r                                                                                  \rIgn:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "Hit:8 https://r2u.stat.illinois.edu/ubuntu jammy Release\n",
            "Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:10 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Get:11 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [2,152 kB]\n",
            "Hit:12 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Get:13 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [2,841 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [2,441 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [51.8 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,422 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [2,954 kB]\n",
            "Get:18 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [921 kB]\n",
            "Fetched 13.0 MB in 2s (6,160 kB/s)\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  liblmdb0 ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  liblmdb0 ncbi-blast+ ncbi-data\n",
            "0 upgraded, 3 newly installed, 0 to remove and 45 not upgraded.\n",
            "Need to get 15.9 MB of archives.\n",
            "After this operation, 71.9 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 liblmdb0 amd64 0.9.24-1build2 [47.6 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncbi-data all 6.1.20170106+dfsg1-9 [3,519 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncbi-blast+ amd64 2.12.0+ds-3build1 [12.3 MB]\n",
            "Fetched 15.9 MB in 1s (28.2 MB/s)\n",
            "Selecting previously unselected package liblmdb0:amd64.\n",
            "(Reading database ... 123594 files and directories currently installed.)\n",
            "Preparing to unpack .../liblmdb0_0.9.24-1build2_amd64.deb ...\n",
            "Unpacking liblmdb0:amd64 (0.9.24-1build2) ...\n",
            "Selecting previously unselected package ncbi-data.\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106+dfsg1-9_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106+dfsg1-9) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.12.0+ds-3build1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.12.0+ds-3build1) ...\n",
            "Setting up liblmdb0:amd64 (0.9.24-1build2) ...\n",
            "Setting up ncbi-data (6.1.20170106+dfsg1-9) ...\n",
            "Setting up ncbi-blast+ (2.12.0+ds-3build1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "\n",
        "# Create a BLAST database\n",
        "command = \"makeblastdb -in sequences.fasta -dbtype nucl\"\n",
        "subprocess.run(command, shell=True, check=True)\n",
        "\n",
        "# Run BLASTN against the database\n",
        "command = \"blastn -query sequences.fasta -db sequences.fasta -out blast_results.xml -outfmt 5\"\n",
        "subprocess.run(command, shell=True, check=True)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iroyeXU_k0fq",
        "outputId": "57bd2d1c-ab93-421b-e3da-d8cfeb031a05"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "CompletedProcess(args='blastn -query sequences.fasta -db sequences.fasta -out blast_results.xml -outfmt 5', returncode=0)"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "from Bio.Blast.Applications import NcbimakeblastdbCommandline, NcbiblastnCommandline\n",
        "\n",
        "# Load the data\n",
        "pdb_data_seq = pd.read_csv('pdb_data_seq.csv')\n",
        "\n",
        "protein_sequences = pdb_data_seq[pdb_data_seq['macromoleculeType'] == 'Protein'].head(10)\n",
        "\n",
        "\n",
        "# Save sequences to a FASTA file\n",
        "with open('sequences.fasta', 'w') as f:\n",
        "    for index, row in protein_sequences.iterrows():\n",
        "        f.write(f\">{row['structureId']}_{row['chainId']}\\n\")\n",
        "        f.write(f\"{row['sequence']}\\n\")\n",
        "\n",
        "# Create a BLAST database\n",
        "makeblastdb_cline = NcbimakeblastdbCommandline(\n",
        "    dbtype=\"nucl\", input_file=\"sequences.fasta\")\n",
        "makeblastdb_cline()\n",
        "\n",
        "# Run BLASTN against the database\n",
        "blastn_cline = NcbiblastnCommandline(\n",
        "    query=\"sequences.fasta\", db=\"sequences.fasta\", outfmt=5, out=\"blast_results.xml\")\n",
        "stdout, stderr = blastn_cline()\n",
        "\n",
        "print(\"Local BLASTN search completed for all sequences.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qxomQmNJkX6m",
        "outputId": "15d341b5-f96b-43ec-923a-0cf7d3f9e5b2"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Local BLASTN search completed for all sequences.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install clustalo\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Jx8VPaGKpNFR",
        "outputId": "f6bfa4e8-d3fa-40cd-e46a-9a5638c1cc69"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libargtable2-0\n",
            "The following NEW packages will be installed:\n",
            "  clustalo libargtable2-0\n",
            "0 upgraded, 2 newly installed, 0 to remove and 45 not upgraded.\n",
            "Need to get 273 kB of archives.\n",
            "After this operation, 694 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libargtable2-0 amd64 13-1.1 [14.1 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 clustalo amd64 1.2.4-7 [259 kB]\n",
            "Fetched 273 kB in 0s (775 kB/s)\n",
            "Selecting previously unselected package libargtable2-0.\n",
            "(Reading database ... 123835 files and directories currently installed.)\n",
            "Preparing to unpack .../libargtable2-0_13-1.1_amd64.deb ...\n",
            "Unpacking libargtable2-0 (13-1.1) ...\n",
            "Selecting previously unselected package clustalo.\n",
            "Preparing to unpack .../clustalo_1.2.4-7_amd64.deb ...\n",
            "Unpacking clustalo (1.2.4-7) ...\n",
            "Setting up libargtable2-0 (13-1.1) ...\n",
            "Setting up clustalo (1.2.4-7) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.4) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Run Clustal Omega in Colab\n",
        "!clustalo -i hydrolase_sequences.fasta -o hydrolase_aligned.aln\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8QYPrk_UpVlw",
        "outputId": "7b726ad5-0d61-4097-f362-9a135a511c1c"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ERROR: Failed to open sequence file hydrolase_sequences.fasta for reading\n",
            "FATAL: Reading sequence file 'hydrolase_sequences.fasta' failed\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "# Load your data\n",
        "merged_data = pd.read_csv('pdb_data_no_dups.csv')\n",
        "seq_data = pd.read_csv('pdb_data_seq.csv')\n",
        "\n",
        "# Filter by a specific classification, e.g., 'HYDROLASE'\n",
        "filtered_data = merged_data[merged_data['classification'] == 'HYDROLASE']\n",
        "filtered_seq_data = seq_data[seq_data['structureId'].isin(filtered_data['structureId'])]\n",
        "\n",
        "# Write filtered sequences to a FASTA file\n",
        "def write_fasta(data, filename):\n",
        "    with open(filename, 'w') as fasta_file:\n",
        "        for index, row in data.iterrows():\n",
        "            fasta_file.write(f\">{row['structureId']}|{row['chainId']}\\n\")\n",
        "            fasta_file.write(f\"{row['sequence']}\\n\")\n",
        "\n",
        "write_fasta(filtered_seq_data, 'hydrolase_sequences.fasta')\n"
      ],
      "metadata": {
        "id": "s9nPlSRwsmTQ"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from tensorflow.keras.models import Sequential\n",
        "from tensorflow.keras.layers import Dense, Conv1D, MaxPooling1D, Flatten\n",
        "from tensorflow.keras.utils import to_categorical\n",
        "\n",
        "# Assuming 'sampled_data' is your preprocessed data\n",
        "# Convert categorical labels to numeric\n",
        "label_encoder = LabelEncoder()\n",
        "merged_data['classification'] = label_encoder.fit_transform(merged_data['classification'])\n",
        "\n",
        "# Split data into features and target\n",
        "X = merged_data[['sequence_length', 'residueCount']]\n",
        "y = to_categorical(merged_data['classification'])\n",
        "\n",
        "# Reshape X to fit CNN input (samples, time steps, features)\n",
        "X = np.expand_dims(X, axis=2)\n",
        "\n",
        "# Train-test split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Build the CNN model\n",
        "cnn_model = Sequential([\n",
        "    Conv1D(64, kernel_size=3, activation='relu', input_shape=(X_train.shape[1], 1)),\n",
        "    MaxPooling1D(pool_size=2),\n",
        "    Flatten(),\n",
        "    Dense(128, activation='relu'),\n",
        "    Dense(y_train.shape[1], activation='softmax')\n",
        "])\n",
        "\n",
        "cnn_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])\n",
        "\n",
        "# Train the model\n",
        "cnn_model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))\n",
        "\n",
        "# Evaluate the model\n",
        "cnn_score = cnn_model.evaluate(X_test, y_test)\n",
        "print(f'CNN Accuracy: {cnn_score[1]}')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 365
        },
        "id": "Nh0JINN0uK0L",
        "outputId": "1ef7dfc2-64aa-4807-c0e6-8760d70f65a4"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "error",
          "ename": "KeyError",
          "evalue": "\"None of [Index(['sequence_length', 'residueCount_x'], dtype='object')] are in the [columns]\"",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-9-cfa06a7dba4c>\u001b[0m in \u001b[0;36m<cell line: 15>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m# Split data into features and target\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmerged_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'residueCount_x'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mto_categorical\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmerged_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'classification'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3897\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3898\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3899\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_indexer_strict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"columns\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3900\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3901\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6113\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6115\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_if_missing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6117\u001b[0m         \u001b[0mkeyarr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6174\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0muse_interval_msg\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6175\u001b[0m                     \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6176\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"None of [{key}] are in the [{axis_name}]\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6177\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6178\u001b[0m             \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmissing_mask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyError\u001b[0m: \"None of [Index(['sequence_length', 'residueCount_x'], dtype='object')] are in the [columns]\""
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pdb_data_no_dups = pd.read_csv('./pdb_data_no_dups.csv')\n",
        "pdb_data_seq = pd.read_csv('./pdb_data_seq.csv')\n",
        "\n",
        "# Merge the datasets on 'structureId'\n",
        "merged_data = pd.merge(pdb_data_no_dups, pdb_data_seq, on='structureId')\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from tensorflow.keras.models import Sequential\n",
        "from tensorflow.keras.layers import Dense, Conv1D, MaxPooling1D, Flatten\n",
        "from tensorflow.keras.utils import to_categorical\n",
        "\n",
        "# Assuming 'merged_data' is your original DataFrame\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "merged_data['sequence'] = merged_data['sequence'].astype(str)\n",
        "merged_data['sequence_length'] = merged_data['sequence'].apply(len)\n",
        "\n",
        "# Convert categorical labels to numeric\n",
        "label_encoder = LabelEncoder()\n",
        "merged_data['classification'] = label_encoder.fit_transform(merged_data['classification'])\n",
        "\n",
        "# Split data into features and target\n",
        "X = merged_data[['sequence_length', 'residueCount']]  # Use the correct column names\n",
        "y = to_categorical(merged_data['classification'])\n",
        "\n",
        "# Reshape X to fit CNN input (samples, time steps, features)\n",
        "X = np.expand_dims(X, axis=2)\n",
        "\n",
        "# Train-test split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Build the CNN model\n",
        "cnn_model = Sequential([\n",
        "    Conv1D(64, kernel_size=3, activation='relu', input_shape=(X_train.shape[1], 1)),\n",
        "    MaxPooling1D(pool_size=2),\n",
        "    Flatten(),\n",
        "    Dense(128, activation='relu'),\n",
        "    Dense(y_train.shape[1], activation='softmax')\n",
        "])\n",
        "\n",
        "cnn_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])\n",
        "\n",
        "# Train the model\n",
        "cnn_model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))\n",
        "\n",
        "# Evaluate the model\n",
        "cnn_score = cnn_model.evaluate(X_test, y_test)\n",
        "print(f'CNN Accuracy: {cnn_score[1]}')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 365
        },
        "id": "7AsDT0xqvyyW",
        "outputId": "4df19b16-c426-4289-ecae-2e0b54302fd8"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "error",
          "ename": "KeyError",
          "evalue": "\"['residueCount'] not in index\"",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-11-23feda433dcf>\u001b[0m in \u001b[0;36m<cell line: 25>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;31m# Split data into features and target\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmerged_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sequence_length'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'residueCount'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m  \u001b[0;31m# Use the correct column names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mto_categorical\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmerged_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'classification'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3897\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3898\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3899\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_indexer_strict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"columns\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3900\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3901\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6113\u001b[0m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_indexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reindex_non_unique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6114\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6115\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_if_missing\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6117\u001b[0m         \u001b[0mkeyarr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36m_raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6177\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6178\u001b[0m             \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mensure_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmissing_mask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6179\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{not_found} not in index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6180\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6181\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0moverload\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyError\u001b[0m: \"['residueCount'] not in index\""
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(merged_data.columns)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c64APZHsyrS2",
        "outputId": "9cdb3d76-9451-4699-8b74-717cf9d27c6a"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Index(['structureId', 'classification', 'experimentalTechnique',\n",
            "       'macromoleculeType_x', 'residueCount_x', 'resolution',\n",
            "       'structureMolecularWeight', 'crystallizationMethod',\n",
            "       'crystallizationTempK', 'densityMatthews', 'densityPercentSol',\n",
            "       'pdbxDetails', 'phValue', 'publicationYear', 'chainId', 'sequence',\n",
            "       'residueCount_y', 'macromoleculeType_y', 'sequence_length'],\n",
            "      dtype='object')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pdb_data_no_dups = pd.read_csv('./pdb_data_no_dups.csv')\n",
        "pdb_data_seq = pd.read_csv('./pdb_data_seq.csv')\n",
        "\n",
        "# Merge the datasets on 'structureId'\n",
        "merged_data = pd.merge(pdb_data_no_dups, pdb_data_seq, on='structureId')\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from tensorflow.keras.models import Sequential\n",
        "from tensorflow.keras.layers import Dense, Conv1D, MaxPooling1D, Flatten\n",
        "from tensorflow.keras.utils import to_categorical\n",
        "\n",
        "# Assuming 'merged_data' is your original DataFrame\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "merged_data['sequence'] = merged_data['sequence'].astype(str)\n",
        "merged_data['sequence_length'] = merged_data['sequence'].apply(len)\n",
        "\n",
        "# Check the column names\n",
        "print(merged_data.columns)\n",
        "\n",
        "# If 'residueCount' does not exist, identify the correct column name\n",
        "# For example, if it's 'residueCount_x' or 'residueCount_y', use that\n",
        "if 'residueCount_x' in merged_data.columns:\n",
        "    residue_count_col = 'residueCount_x'\n",
        "elif 'residueCount_y' in merged_data.columns:\n",
        "    residue_count_col = 'residueCount_y'\n",
        "else:\n",
        "    residue_count_col = 'residueCount'  # Adjust based on your dataset\n",
        "\n",
        "# Convert categorical labels to numeric\n",
        "label_encoder = LabelEncoder()\n",
        "merged_data['classification'] = label_encoder.fit_transform(merged_data['classification'])\n",
        "# Sample a smaller portion of the data, e.g., 10% of the dataset\n",
        "merged_data = merged_data.sample(frac=0.1, random_state=42)\n",
        "\n",
        "# Split data into features and target\n",
        "X = merged_data[['sequence_length', residue_count_col]]  # Use the correct column name\n",
        "y = to_categorical(merged_data['classification'])\n",
        "\n",
        "# Reshape X to fit CNN input (samples, time steps, features)\n",
        "X = np.expand_dims(X, axis=2)\n",
        "\n",
        "# Train-test split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Build the CNN model\n",
        "cnn_model = Sequential([\n",
        "    Conv1D(64, kernel_size=3, activation='relu', input_shape=(X_train.shape[1], 1)),\n",
        "    MaxPooling1D(pool_size=2),\n",
        "    Flatten(),\n",
        "    Dense(128, activation='relu'),\n",
        "    Dense(y_train.shape[1], activation='softmax')\n",
        "])\n",
        "\n",
        "cnn_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])\n",
        "\n",
        "# Train the model\n",
        "cnn_model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))\n",
        "\n",
        "# Evaluate the model\n",
        "cnn_score = cnn_model.evaluate(X_test, y_test)\n",
        "print(f'CNN Accuracy: {cnn_score[1]}')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 706
        },
        "id": "fT5-fVTeyu90",
        "outputId": "f6fdeb08-f8c6-4c13-e60c-99d0ec9fd683"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Index(['structureId', 'classification', 'experimentalTechnique',\n",
            "       'macromoleculeType_x', 'residueCount_x', 'resolution',\n",
            "       'structureMolecularWeight', 'crystallizationMethod',\n",
            "       'crystallizationTempK', 'densityMatthews', 'densityPercentSol',\n",
            "       'pdbxDetails', 'phValue', 'publicationYear', 'chainId', 'sequence',\n",
            "       'residueCount_y', 'macromoleculeType_y', 'sequence_length'],\n",
            "      dtype='object')\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.\n",
            "  super().__init__(activity_regularizer=activity_regularizer, **kwargs)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Epoch 1/10\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "Exception encountered when calling Conv1D.call().\n\n\u001b[1mNegative dimension size caused by subtracting 3 from 2 for '{{node sequential_1/conv1d_1/convolution}} = Conv2D[T=DT_FLOAT, data_format=\"NHWC\", dilations=[1, 1, 1, 1], explicit_paddings=[], padding=\"VALID\", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true](sequential_1/conv1d_1/convolution/ExpandDims, sequential_1/conv1d_1/convolution/ExpandDims_1)' with input shapes: [?,1,2,1], [1,3,1,64].\u001b[0m\n\nArguments received by Conv1D.call():\n  • inputs=tf.Tensor(shape=(None, 2, 1), dtype=float32)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-2-eb78e2e640a7>\u001b[0m in \u001b[0;36m<cell line: 59>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[0;31m# Train the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 59\u001b[0;31m \u001b[0mcnn_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     60\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m \u001b[0;31m# Evaluate the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    120\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m             \u001b[0;31m# `keras.config.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 122\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    120\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m             \u001b[0;31m# `keras.config.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 122\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: Exception encountered when calling Conv1D.call().\n\n\u001b[1mNegative dimension size caused by subtracting 3 from 2 for '{{node sequential_1/conv1d_1/convolution}} = Conv2D[T=DT_FLOAT, data_format=\"NHWC\", dilations=[1, 1, 1, 1], explicit_paddings=[], padding=\"VALID\", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true](sequential_1/conv1d_1/convolution/ExpandDims, sequential_1/conv1d_1/convolution/ExpandDims_1)' with input shapes: [?,1,2,1], [1,3,1,64].\u001b[0m\n\nArguments received by Conv1D.call():\n  • inputs=tf.Tensor(shape=(None, 2, 1), dtype=float32)"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pdb_data_no_dups = pd.read_csv('./pdb_data_no_dups.csv')\n",
        "pdb_data_seq = pd.read_csv('./pdb_data_seq.csv')\n",
        "\n",
        "# Merge the datasets on 'structureId'\n",
        "merged_data = pd.merge(pdb_data_no_dups, pdb_data_seq, on='structureId')\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from tensorflow.keras.models import Sequential\n",
        "from tensorflow.keras.layers import Dense, Conv1D, MaxPooling1D, Flatten\n",
        "from tensorflow.keras.utils import to_categorical\n",
        "\n",
        "# Assuming 'merged_data' is your original DataFrame\n",
        "# Ensure 'sequence' column is of type str and create 'sequence_length'\n",
        "merged_data['sequence'] = merged_data['sequence'].astype(str)\n",
        "merged_data['sequence_length'] = merged_data['sequence'].apply(len)\n",
        "\n",
        "# Convert categorical labels to numeric\n",
        "label_encoder = LabelEncoder()\n",
        "merged_data['classification'] = label_encoder.fit_transform(merged_data['classification'])\n",
        "\n",
        "# Use a smaller sample of data if the dataset is too large\n",
        "sampled_data = merged_data.sample(frac=0.1, random_state=42)\n",
        "\n",
        "# Use the correct column for residue count\n",
        "residue_count_col = 'residueCount_x' if 'residueCount_x' in merged_data.columns else 'residueCount_y'\n",
        "\n",
        "# Split data into features and target\n",
        "X = sampled_data[['sequence_length', residue_count_col]]\n",
        "y = to_categorical(sampled_data['classification'])\n",
        "\n",
        "# Reshape X to fit CNN input (samples, time steps, features)\n",
        "X = np.expand_dims(X, axis=2)\n",
        "\n",
        "# Train-test split\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
        "\n",
        "# Build the CNN model with a smaller kernel size or same padding\n",
        "cnn_model = Sequential([\n",
        "    Conv1D(64, kernel_size=2, activation='relu', input_shape=(X_train.shape[1], 1)),  # Adjusted kernel size\n",
        "    MaxPooling1D(pool_size=2),\n",
        "    Flatten(),\n",
        "    Dense(128, activation='relu'),\n",
        "    Dense(y_train.shape[1], activation='softmax')\n",
        "])\n",
        "\n",
        "cnn_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])\n",
        "\n",
        "# Train the model\n",
        "cnn_model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))\n",
        "\n",
        "# Evaluate the model\n",
        "cnn_score = cnn_model.evaluate(X_test, y_test)\n",
        "print(f'CNN Accuracy: {cnn_score[1]}')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 563
        },
        "id": "teXc4bLZ1ha2",
        "outputId": "1f104010-c0a5-43e9-a7a7-c4a03a700ea9"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.\n",
            "  super().__init__(activity_regularizer=activity_regularizer, **kwargs)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Epoch 1/10\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "Exception encountered when calling MaxPooling1D.call().\n\n\u001b[1mNegative dimension size caused by subtracting 2 from 1 for '{{node sequential_1_1/max_pooling1d_1_1/MaxPool1d}} = MaxPool[T=DT_FLOAT, data_format=\"NHWC\", explicit_paddings=[], ksize=[1, 1, 2, 1], padding=\"VALID\", strides=[1, 1, 2, 1]](sequential_1_1/max_pooling1d_1_1/MaxPool1d/ExpandDims)' with input shapes: [?,1,1,64].\u001b[0m\n\nArguments received by MaxPooling1D.call():\n  • inputs=tf.Tensor(shape=(None, 1, 64), dtype=float32)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-3-2b1c6cfbdadd>\u001b[0m in \u001b[0;36m<cell line: 52>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m \u001b[0;31m# Train the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m \u001b[0mcnn_model\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     54\u001b[0m \u001b[0;31m# Evaluate the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    120\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m             \u001b[0;31m# `keras.config.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 122\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    120\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m             \u001b[0;31m# `keras.config.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 122\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: Exception encountered when calling MaxPooling1D.call().\n\n\u001b[1mNegative dimension size caused by subtracting 2 from 1 for '{{node sequential_1_1/max_pooling1d_1_1/MaxPool1d}} = MaxPool[T=DT_FLOAT, data_format=\"NHWC\", explicit_paddings=[], ksize=[1, 1, 2, 1], padding=\"VALID\", strides=[1, 1, 2, 1]](sequential_1_1/max_pooling1d_1_1/MaxPool1d/ExpandDims)' with input shapes: [?,1,1,64].\u001b[0m\n\nArguments received by MaxPooling1D.call():\n  • inputs=tf.Tensor(shape=(None, 1, 64), dtype=float32)"
          ]
        }
      ]
    }
  ]
}